{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: here\n",
      "\n",
      "Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :\n",
      "“there is no package called ‘here’”\n",
      "Loading required package: data.table\n",
      "\n",
      "Loading required package: tidyverse\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.5     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.3     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.7\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m    masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m       masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m      masks \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\n",
      "Loading required package: reshape2\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "Loading required package: ggpubr\n",
      "\n",
      "Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :\n",
      "“there is no package called ‘ggpubr’”\n",
      "Loading required package: cowplot\n",
      "\n",
      "Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :\n",
      "“there is no package called ‘cowplot’”\n",
      "Loading required package: coin\n",
      "\n",
      "Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :\n",
      "“there is no package called ‘coin’”\n"
     ]
    }
   ],
   "source": [
    "knitr::opts_chunk$set(echo = TRUE)\n",
    "require(here)\n",
    "require(data.table)\n",
    "require(tidyverse)\n",
    "require(reshape2)\n",
    "require(ggpubr)\n",
    "require(cowplot)\n",
    "require(coin)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(echo = TRUE)\n",
    "dd_genes <- fread(here('data/discordant_consensus_novel_DD_associated_genes_kaplanis2020.txt'),header=FALSE)[,V1]\n",
    "hotspot_genes <- fread(here('data/hotspot_genes.txt'),header=FALSE)[,V1]\n",
    "novel_hotspot_genes <- fread(here('data/proposed_novel_hotspot_genes.txt'),header=FALSE)[,V1]\n",
    "ion_channel_genes <- fread(here('data/All_genes_with_PF00520_domain.txt'),header=FALSE)[,V1]\n",
    "hotspot_genes_lenient <- fread(here('data/lenient_hotspot_genes.txt'),header=FALSE)[,V1]\n",
    "gene_tpm <- fread(here('data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz'),header=TRUE)\n",
    "constraint_in <- fread(here('data/gnomad.v2.1.1.lof_metrics.by_gene.txt'),header=TRUE)\n",
    "mpc_in <- fread(here('data/mpc_missense_constraint_multiple_regions_samocha17.txt'),header=TRUE)\n",
    "regional_constraint <- fread(here('data/hotspot_mutation_regional_constraint.txt'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Input data and counts:**\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove novel hotspot genes from hotspot gene list\n",
    "hotspot_genes <- hotspot_genes[!hotspot_genes %in% novel_hotspot_genes]\n",
    "\n",
    "#Remove hotspot genes from DD consensus gene list\n",
    "dd_genes <- dd_genes[!dd_genes %in% hotspot_genes]\n",
    "\n",
    "#Parse gene TPM file\n",
    "gene_tpm <- gene_tpm[,Gene_ID:=gsub(\"\\\\..*\",\"\",Name)]\n",
    "gene_tpm <- melt(gene_tpm,id.vars = c('Description','Name','Gene_ID'))\n",
    "colnames(gene_tpm) <- c('hgnc','ensembl_versioned','ensembl','tissue','tpm')\n",
    "\n",
    "gene_tpm <- gene_tpm[,class:=ifelse(hgnc %in% novel_hotspot_genes,'Proposed Novel Hotspot Genes',\n",
    "                                    ifelse(hgnc %in% hotspot_genes,'Hotspot Genes',\n",
    "                                           ifelse(hgnc %in% dd_genes,'DD Consensus Genes','Control Genes')))]\n",
    "\n",
    "gene_tpm <- gene_tpm[,contains_ion_channel:=ifelse(hgnc %in% ion_channel_genes,TRUE,FALSE)]\n",
    "\n",
    "head(gene_tpm)\n",
    "\n",
    "gene_tpm_tab <- gene_tpm %>%\n",
    "  group_by(class) %>%\n",
    "  dplyr::select('hgnc','class') %>%\n",
    "  unique() %>%\n",
    "  count()\n",
    "\n",
    "colnames(gene_tpm_tab) <- c('Class','Count')\n",
    "  \n",
    "#Table of counts by class (Control Genes, DD Consensus Genes, Hotspot Genes, Proposed Novel Hotspot Genes)\n",
    "knitr::kable(gene_tpm_tab)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Counts post-sampling:**\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set seed\n",
    "set.seed(1255)\n",
    "\n",
    "sampled_genes <- sample(unique(gene_tpm[class=='Control Genes',hgnc]),size=1000,replace=FALSE)\n",
    "\n",
    "gene_tpm_sampled <- gene_tpm[class!='Control Genes' | hgnc %in% sampled_genes]\n",
    "\n",
    "gene_tpm_sampled_tab <- gene_tpm_sampled %>%\n",
    "  group_by(class) %>%\n",
    "  dplyr::select('hgnc','class') %>%\n",
    "  unique() %>%\n",
    "  count()\n",
    "\n",
    "colnames(gene_tpm_sampled_tab) <- c('Class','Count')\n",
    "\n",
    "#Table of sampled counts by class (Control Genes, DD Consensus Genes, Hotspot Genes, Proposed Novel Hotspot Genes)\n",
    "knitr::kable(gene_tpm_sampled_tab)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4A: Hotspot genes share a characteristic gene expression pattern across tissues\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set seed\n",
    "set.seed(1255)\n",
    "\n",
    "#Get significant differences between groups by tissue; Bonferroni correct for multiple testing\n",
    "getProportionP <- function( class1, class2 ) {\n",
    "  \n",
    "  temp <- gene_tpm_sampled[class==class1 | class==class2]\n",
    "  temp <- temp[,expressed:=ifelse(tpm > 1, TRUE, FALSE)]\n",
    "  temp <- temp[,c('hgnc','class','tissue','expressed')]\n",
    "  \n",
    "  fisherTest <- function( t ) {\n",
    "    \n",
    "    class1_temp <- temp[class==class1 & tissue==t]\n",
    "    expressed_class1 <- nrow(class1_temp[which(expressed==TRUE)])\n",
    "    unexpressed_class1 <- nrow(class1_temp[which(expressed==FALSE)])\n",
    "\n",
    "    class2_temp <- temp[class==class2 & tissue==t]\n",
    "    expressed_class2 <- nrow(class2_temp[which(expressed==TRUE)])\n",
    "    unexpressed_class2 <- nrow(class2_temp[which(expressed==FALSE)])\n",
    "  \n",
    "    fisher_p <- fisher.test(cbind(c(expressed_class1,unexpressed_class1),\n",
    "                    c(expressed_class2,unexpressed_class2)))$p.value\n",
    "  \n",
    "    return(fisher_p)\n",
    "    \n",
    "  }\n",
    "  \n",
    "  temp <- temp[,fisher_p:=fisherTest(tissue),by=tissue]\n",
    "  temp <- unique(temp[,c('tissue','fisher_p')])\n",
    "  temp <- temp[,fisher_corrected_p:=p.adjust(fisher_p,method='bonferroni')]\n",
    "  return(temp)\n",
    "  \n",
    "}\n",
    "\n",
    "getPermutationP <- function( class1, class2 ) {\n",
    "  \n",
    "  temp <- gene_tpm_sampled[ (class==class1 | class==class2)]\n",
    "  temp <- temp[,c('class','tissue','tpm')]\n",
    "  temp <- temp[,permutation_p:=pvalue(independence_test(tpm ~ factor(class))),by=tissue]\n",
    "  temp <- unique(temp[,c('tissue','permutation_p')])\n",
    "  temp <- temp[,permutation_corrected_p:=p.adjust(permutation_p,method='bonferroni')]\n",
    "  return(temp)\n",
    "  \n",
    "}\n",
    "\n",
    "dd_consensus_control_prop_table <- getProportionP('Hotspot Genes','DD Consensus Genes')\n",
    "dd_consensus_control_tpm_table <- getPermutationP('DD Consensus Genes','Control Genes')\n",
    "knitr::kable(dd_consensus_control_prop_table)\n",
    "knitr::kable(dd_consensus_control_tpm_table)\n",
    "\n",
    "#Randomly sample DD Consensus and Control Genes\n",
    "sampleGenes <- function( iterations, sampling_size ) {\n",
    "\n",
    "  out <- vector('list',length=iterations)\n",
    "  \n",
    "  for( i in 1:iterations ) {\n",
    "    \n",
    "    sampled_dat <- gene_tpm_sampled %>% \n",
    "    filter(class!='Hotspot Genes' & class!='Proposed Novel Hotspot Genes') %>%\n",
    "    group_by(tissue,class) %>% \n",
    "    sample_n(sampling_size,replace=TRUE) %>%\n",
    "    mutate(i=factor(i))\n",
    "    \n",
    "    out[[i]] <- data.table(sampled_dat)\n",
    "    \n",
    "  }\n",
    "  \n",
    "  out_dt <- do.call(rbind,out)\n",
    "  return(out_dt)\n",
    "\n",
    "}\n",
    "\n",
    "hotspot_tpm <- gene_tpm_sampled %>%\n",
    "  filter(class=='Hotspot Genes') %>%\n",
    "  mutate(i=0)\n",
    "\n",
    "#Bind sampled genes to hotspot genes; calculate the proportion expressed, mean, and standard deviation across sets\n",
    "fig_a <- rbind(sampleGenes(1000,25),hotspot_tpm) %>%\n",
    "  group_by(tissue,class,i) %>%\n",
    "  summarise(\n",
    "    prop_expressed = sum(tpm > 1, na.rm=T) / n(),\n",
    "    .groups = 'drop_last'\n",
    "  ) %>%\n",
    "  ungroup() %>%\n",
    "  group_by(tissue,class) %>%\n",
    "  summarise(\n",
    "    mean = mean(prop_expressed),\n",
    "    min = mean - sd(prop_expressed),\n",
    "    max = mean + sd(prop_expressed),\n",
    "    .groups = 'keep'\n",
    "  ) %>%\n",
    "  data.table()\n",
    "\n",
    "#Figure 4A\n",
    "a_fill <- c('Control Genes' = 'skyblue3',\n",
    "            'DD Consensus Genes' = 'darkgreen',\n",
    "            'Hotspot Genes' = 'darkgoldenrod')\n",
    "\n",
    "a <- ggplot(fig_a[class!='Proposed Novel Hotspot Genes' & class!='Hotspot Genes'],aes(x=tissue,y=mean,colour=class,ymin=min,ymax=max)) +\n",
    "  geom_point(alpha=0.7,size=3,shape=15) +\n",
    "  geom_linerange(alpha=0.5,size=3) + \n",
    "  geom_point(data=fig_a[class=='Hotspot Genes'],aes(x=tissue,y=mean,colour=class),size=5) +\n",
    "  theme_pubr() +\n",
    "  theme(axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=16), \n",
    "        axis.text.x = element_text(angle = 90, \n",
    "                                   vjust = 0.5, \n",
    "                                   hjust=1),\n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_colour_manual(values=a_fill) +\n",
    "  ylim(0,1) +\n",
    "  xlab('') +\n",
    "  ylab('Proportion expressed genes (TPM > 1)')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4BC: Distinct expression pattern in hotspot genes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set seed\n",
    "set.seed(1255)\n",
    "\n",
    "brain_tissues <- c('Brain - Amygdala','Brain - Anterior cingulate cortex (BA24)','Brain - Caudate (basal ganglia)','Brain - Cerebellar Hemisphere','Brain - Cerebellum',\n",
    "                   'Brain - Cortex','Brain - Frontal Cortex (BA9)','Brain - Hippocampus','Brain - Hypothalamus','Brain - Nucleus accumbens (basal ganglia)','Brain - Putamen (basal ganglia)',\n",
    "                   'Brain - Spinal cord (cervical c-1)','Brain - Substantia nigra')\n",
    "\n",
    "#Subset gene_tpm to randomly sampled genes, but additionally include all genes that contain ion channels.\n",
    "#Group by tissue group (Brain, Other Tissues) and calculate the median TPM per tissue group.\n",
    "median_gene_tpm <- gene_tpm %>%\n",
    "  filter(class!='Control Genes' | hgnc %in% sampled_genes | contains_ion_channel==TRUE) %>%\n",
    "  mutate(tissue_group = ifelse(tissue %in% brain_tissues,'Brain','Other Tissues')) %>%\n",
    "  group_by(hgnc) %>%\n",
    "  mutate(wilcox_p = wilcox.test(tpm ~ tissue_group,exact=FALSE)$p.value,\n",
    "         wilcox_corrected_p = p.adjust(wilcox_p, method = 'bonferroni')) %>%\n",
    "  group_by(hgnc,tissue_group) %>%\n",
    "  mutate(median_tpm_tissuegroup = median(tpm)) %>%\n",
    "  select(-c('tpm','tissue','ensembl_versioned')) %>%\n",
    "  unique() %>%\n",
    "  reshape2::dcast(formula=hgnc + class + contains_ion_channel + wilcox_p + wilcox_corrected_p ~ tissue_group, value.var='median_tpm_tissuegroup') %>%\n",
    "  data.table()\n",
    "\n",
    "#Calculate the proportion expressed higher in brain\n",
    "fig_b <- median_gene_tpm %>%\n",
    "  group_by(class) %>%\n",
    "  summarise(\n",
    "    num_brain = sum(Brain > `Other Tissues`, na.rm=T),\n",
    "    num_other_tissues = sum(`Other Tissues` > Brain, na.rm=T),\n",
    "    total = n(),\n",
    "    num_unexpressed = total - (num_brain + num_other_tissues),\n",
    "    prop_brain = sum(Brain > `Other Tissues`, na.rm=T) / n(),\n",
    "    .groups='drop_last'\n",
    "  ) %>%\n",
    "  reshape2::melt(id.vars=c('class')) %>%\n",
    "  data.table()\n",
    "\n",
    "knitr::kable(fig_b)\n",
    "\n",
    "#Fisher's exact: Hotspot genes versus control/DD consensus genes\n",
    "hotspot_counts <- c(fig_b[class=='Hotspot Genes' & variable=='num_brain',value],\n",
    "                    fig_b[class=='Hotspot Genes' & variable=='num_other_tissues',value])\n",
    "dd_counts <- c(fig_b[class=='DD Consensus Genes' & variable=='num_brain',value],\n",
    "                    fig_b[class=='DD Consensus Genes' & variable=='num_other_tissues',value])\n",
    "control_counts <- c(fig_b[class=='Control Genes' & variable=='num_brain',value],\n",
    "                    fig_b[class=='Control Genes' & variable=='num_other_tissues',value])\n",
    "\n",
    "fisher.test(cbind(hotspot_counts,control_counts))\n",
    "fisher.test(cbind(hotspot_counts,dd_counts))\n",
    "\n",
    "#Calculate TPM differences between classes\n",
    "fig_c <- median_gene_tpm %>%\n",
    "  mutate(diff = Brain - `Other Tissues`) %>%\n",
    "  mutate(class = ifelse(hgnc %in% ion_channel_genes & class=='DD Consensus Genes','PF00520-Containing DD Consensus Genes',\n",
    "                        ifelse(hgnc %in% ion_channel_genes & class=='Control Genes','PF00520-Containing Control Genes',class)))\n",
    "\n",
    "fig_c_supp <- gene_tpm %>%\n",
    "  filter(class!='Control Genes' | hgnc %in% sampled_genes | contains_ion_channel==TRUE) %>%\n",
    "  mutate(tissue_group = ifelse(tissue %in% brain_tissues,'Brain','Other Tissues')) %>%\n",
    "  mutate(class = ifelse(hgnc %in% ion_channel_genes & class=='DD Consensus Genes','PF00520-Containing DD Consensus Genes',\n",
    "                        ifelse(hgnc %in% ion_channel_genes & class=='Control Genes','PF00520-Containing Control Genes',class))) %>%\n",
    "  filter(class %in% c('PF00520-Containing Control Genes','PF00520-Containing DD Consensus Genes','Hotspot Genes')) %>%\n",
    "  mutate(class = factor(class, levels = c('PF00520-Containing Control Genes','PF00520-Containing DD Consensus Genes','Hotspot Genes')))\n",
    "\n",
    "ggplot(fig_c_supp,aes(x=tissue,y=tpm,fill=class)) + \n",
    "  geom_boxplot(outlier.colour = NA, alpha = 0.8) +\n",
    "  theme_pubr() +\n",
    "  theme(axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=16), \n",
    "        axis.text.x = element_text(angle = 90, \n",
    "                                   vjust = 0.5, \n",
    "                                   hjust=1),\n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  ylim(0,25) +\n",
    "  xlab('') +\n",
    "  ylab('TPM by Tissue') +\n",
    "  scale_fill_manual(values=c('skyblue4','darkgreen','darkgoldenrod'))\n",
    "\n",
    "ggplot(fig_c_supp[tpm > 1],aes(x=tissue,y=tpm,fill=class)) + \n",
    "  geom_boxplot(outlier.colour = NA, alpha = 0.8) +\n",
    "  theme_pubr() +\n",
    "  theme(axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=16), \n",
    "        axis.text.x = element_text(angle = 90, \n",
    "                                   vjust = 0.5, \n",
    "                                   hjust=1),\n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  ylim(0,25) +\n",
    "  xlab('') +\n",
    "  ylab('TPM by Tissue (TPM > 1)') +\n",
    "  scale_fill_manual(values=c('skyblue4','darkgreen','darkgoldenrod'))\n",
    "\n",
    "#Figure 4B\n",
    "b_x_labels = c('Control\\nGenes','DD Consensus\\nGenes','Hotspot\\nGenes','Proposed Novel\\nHotspot Genes')\n",
    "b_levels = c('prop_unexpressed','prop_other','prop_brain')\n",
    "b_fill = c('prop_unexpressed' = 'darkgrey',\n",
    "         'prop_other' = 'steelblue3',\n",
    "         'prop_brain' = 'steelblue4')\n",
    "b_fill_labels = c('prop_unexpressed' = 'Unexpressed',\n",
    "                'prop_other' = 'Other Tissues TPM > Brain TPM',\n",
    "                'prop_brain' = 'Brain TPM > Other Tissues TPM')\n",
    "\n",
    "b <- ggplot(fig_b[variable=='prop_brain'],aes(x=class,y=value,fill=factor(variable,levels=b_levels))) + \n",
    "  geom_col(colour='black',alpha=0.8) +\n",
    "  theme_pubr() +\n",
    "  theme(legend.position = 'none',\n",
    "        axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=14), \n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_fill_manual(labels=b_fill_labels,values=b_fill) +\n",
    "  scale_x_discrete(labels=b_x_labels) +\n",
    "  ylim(0,1) +\n",
    "  xlab('') +\n",
    "  ylab('Proportion of Genes')\n",
    "\n",
    "#Figure 4C\n",
    "c_levels = c('PF00520-Containing Control Genes','PF00520-Containing DD Consensus Genes','Hotspot Genes','Proposed Novel Hotspot Genes')\n",
    "c_labels = c('PF00520-Containing Control Genes','PF00520-Containing DD Consensus Genes','Hotspot Genes')\n",
    "c_x_labels = c('PF00520-Containing\\nControl Genes','PF00520-Containing\\nDD Consensus Genes','Hotspot Genes','Proposed Novel\\nHotspot Genes')\n",
    "c_fill = c('PF00520-Containing Control Genes' = 'skyblue3',\n",
    "         'PF00520-Containing DD Consensus Genes' = 'darkgreen',\n",
    "         'Hotspot Genes' = 'darkgoldenrod')\n",
    "\n",
    "#Artificially sets the diff of TPCN1 for plotting to -10; true value -19.88\n",
    "fig_c <- fig_c[,diff:=ifelse(class=='Proposed Novel Hotspot Genes' & diff < -10,-10,diff)]\n",
    "  \n",
    "c <- ggplot(fig_c[class!='Control Genes' & class!='DD Consensus Genes' & class!='Proposed Novel Hotspot Genes'],\n",
    "              aes(x=factor(class,levels=c_levels),y=diff,colour=factor(class,levels=c_labels))) + \n",
    "  geom_boxplot(alpha=0.8, outlier.shape = NA) + \n",
    "  geom_point(data=fig_c[class=='Proposed Novel Hotspot Genes'],aes(x=3,y=diff),\n",
    "             colour='black',\n",
    "             shape=4,\n",
    "             size=5,\n",
    "             position = position_jitter(0.2)) +\n",
    "  theme_pubr() +\n",
    "  theme(legend.position = 'none',\n",
    "        axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=14), \n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_colour_manual(values=c_fill) +\n",
    "  scale_x_discrete(labels=c_x_labels) +\n",
    "  coord_cartesian(ylim=c(-10,20)) +\n",
    "  ggsignif::geom_signif(\n",
    "    comparisons = list(c('PF00520-Containing Control Genes','PF00520-Containing DD Consensus Genes'),c('PF00520-Containing DD Consensus Genes','Hotspot Genes')),\n",
    "    map_signif_level = FALSE,\n",
    "    y_position = c(12,17),\n",
    "    colour='black') +\n",
    "  xlab('') +\n",
    "  ylab('TPM Brain - TPM Other Tissues')\n",
    "\n",
    "#Figure 4\n",
    "plots <- align_plots(a, b, align = 'v', axis = 'l')\n",
    "bottom_row <- plot_grid(plots[[2]], c, labels = c('B', 'C'), label_size = 14, rel_widths = c(1,1.55))\n",
    "plot_grid(plots[[1]], bottom_row, labels = c('A', ''), label_size = 14, ncol = 1, rel_heights = c(1.75,1))\n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Figure 4. Hotspot genes have a distinct gene expression pattern.** *A) Hotspot genes share a characteristic gene expression pattern across tissues.* Expression of the 19 established DD genes containing a mutation hotspot (hotspot genes, yellow) was evaluated across 54 GTEx tissues. Hotspot gene expression was compared to DD genes (DD consensus, discordant, and novel genes, blue; n = 991) and a randomly sampled set of 1000 GENCODE v37 genes not associated with intellectual disability or developmental delay (control genes, blue). The proportion of expressed genes was calculated based on the median transcripts per million (TPM) per gene in each of the 54 tissues; TPM > 1 was considered expressed. For appropriate comparison to hotspot genes, DD and control genes were randomly sampled into sets of 25 (1000 iterations) and the proportion of expressed genes was computed per set. The mean (square) and standard deviation (bars) of the distribution across these sets is shown in Figure 4A. *B) Hotspot genes are highly expressed in brain.* Given the distinct profile of hotspot gene expression between brain tissues and most other GTEx tissues, we subset the data into two tissue groups (brain and other tissues; see Methods). The median TPM per gene was calculated for each tissue group, and these median TPMs were compared to determine the proportion of the control, DD, hotspot, and proposed novel hotspot gene sets with higher expression in brain than whole blood (dark blue). The remaining genes in each set either have higher expression in other tissues than brain (light blue; Supplementary Figure N1) or are constitutively unexpressed (grey, Supplementary Figure N1). Hotspot genes have a significantly higher proportion of genes with expression in brain > other tissues when compared with control genes (p = 2.985 x 10-5, Fisher’s exact test) and DD consensus genes (p = 0.002). Of the proposed novel hotspot genes, 3 (*CACNA1B*, *KCNH5*, and *KCNG1*) share this property. *C) Hotspot genes are overexpressed in brain compared to other DD ion channel-containing genes.* Because all identified hotspot genes contain a PF00520 ion channel domain, we tested whether the difference between brain TPM and other tissue TPM (differential expression) could differentiate hotspot genes from other PF00520 domain-containing genes. Positive values on the y-axis indicate higher expression in brain, and negative values indicate higher expression in other tissues. PF00520-containing DD genes have significantly higher differential expression compared to PF00520-containing control genes (p = 0.019, Wilcoxon rank-sum test). Hotspot genes have significantly higher differential expression when compared to other PF00520-containing DD genes (p = 0.039, Wilcoxon rank-sum test). Black crosses on the hotspot distribution indicate differential expression of named proposed novel hotspot genes (n = 6) between tissue groups.\n",
    "\n",
    "### Figure 5: Constraint\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get relevant constraint information\n",
    "probabilistic_constraint <- c('pLI','pRec','pNull')\n",
    "upper_bound_fraction <- c('oe_lof_upper','oe_mis_upper','oe_syn_upper')\n",
    "observed_expected <- c('oe_lof','oe_mis','oe_syn','oe_lof_lower','oe_mis_lower','oe_syn_lower',upper_bound_fraction)\n",
    "z_scores <- c('lof_z','mis_z','syn_z')\n",
    "all_constraint <- unique(c(probabilistic_constraint,upper_bound_fraction,observed_expected,z_scores))\n",
    "\n",
    "constraint <- constraint_in %>%\n",
    "  subset(select=c('gene',all_constraint)) %>%\n",
    "  reshape2::melt(id.vars = 'gene', value.name = 'value', var.name = 'constraint_metric') %>%\n",
    "  filter(!is.na(value)) %>%\n",
    "  mutate(class=ifelse(gene %in% novel_hotspot_genes,'Proposed Novel Hotspot Genes',\n",
    "                      ifelse(gene %in% hotspot_genes,'Hotspot Genes',\n",
    "                             ifelse(gene %in% dd_genes,'DD Consensus Genes','Control Genes'))))\n",
    "\n",
    "#Constraint counts\n",
    "constraint_table <- constraint %>%\n",
    "  select(c('gene','class')) %>%\n",
    "  unique() %>%\n",
    "  group_by(class) %>%\n",
    "  count()\n",
    "\n",
    "colnames(constraint_table) <- c('Class','Count')\n",
    "\n",
    "knitr::kable(constraint_table)\n",
    "\n",
    "constraint_table_2 <- constraint_in %>%\n",
    "  subset(select=c('gene',all_constraint)) %>%\n",
    "  filter(gene %in% hotspot_genes | gene %in% novel_hotspot_genes)\n",
    "\n",
    "#Figure 5A\n",
    "fig_5a <- constraint %>%\n",
    "  filter(variable %in% observed_expected) %>%\n",
    "  group_by(class,variable) %>%\n",
    "  mutate(variable=gsub('oe_','',variable)) %>%\n",
    "  mutate(type=substr(variable,1,3)) %>%\n",
    "  mutate(bound=ifelse(grepl('upper',variable),'upper',\n",
    "                      ifelse(grepl('lower',variable),'lower',''))) %>%\n",
    "  group_by(class,type,bound) %>%\n",
    "  summarise(\n",
    "    mean = mean(value),\n",
    "    .groups = 'drop_last'\n",
    "  ) %>%\n",
    "  mutate(bound=ifelse(bound=='','oe',bound)) %>%\n",
    "  reshape2::dcast(formula=class + type ~ bound, value.var='mean')\n",
    "\n",
    "a_labels = c('Loss of function','Missense','Synonymous')\n",
    "a_levels = c('Control Genes','DD Consensus Genes','Hotspot Genes','Proposed Novel Hotspot Genes')\n",
    "a_x_labels = c('Control Genes','DD Consensus Genes','Hotspot Genes','Proposed Novel\\nHotspot Genes')\n",
    "\n",
    "a <- ggplot(fig_5a,aes(x=factor(class,levels=a_levels),y=oe,colour=type,ymin=lower,ymax=upper)) + \n",
    "  geom_pointrange(position=position_dodge(width=1),size=1.6) + \n",
    "  geom_hline(yintercept=1,linetype='dashed',colour='black') +\n",
    "  theme_pubr() +\n",
    "  theme(axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=16), \n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_x_discrete(labels=a_x_labels) +\n",
    "  scale_colour_manual(labels=a_labels, values=c('tomato4','darkorange3','deeppink4')) + \n",
    "  ylab('Observed/expected counts') +\n",
    "  xlab('')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get global missense constraint\n",
    "global_oe_mis <- constraint %>%\n",
    "  filter(variable=='oe_mis')\n",
    "\n",
    "#Add global missense constraint to regional constraint infile and parse\n",
    "fig_5b <- regional_constraint %>%\n",
    "  inner_join(global_oe_mis,by='gene') %>%\n",
    "  mutate(family=ifelse(grepl('CACNA1',gene),'CACNA1\\nGene Family',\n",
    "                             ifelse(grepl('KCN',gene),'KCN\\nGene Family',\n",
    "                                    ifelse(grepl('SCN',gene),'SCN\\nGene Family','Other')))) %>%\n",
    "  dplyr::select(c('obs_exp','value','gene','family')) %>%\n",
    "  unique() %>%\n",
    "  data.table()\n",
    "\n",
    "#Figure 5B\n",
    "b_levels = c('CACNA1\\nGene Family','KCN\\nGene Family','SCN\\nGene Family','Other')\n",
    "b_colour = c('CACNA1\\nGene Family' = 'darkolivegreen',\n",
    "             'KCN\\nGene Family' = 'bisque2',\n",
    "             'SCN\\nGene Family' = 'lightpink4',\n",
    "             'Other' = 'chocolate4')\n",
    "\n",
    "b <- ggplot(fig_5b,aes(x=value,y=obs_exp,colour=factor(family,levels=b_levels))) + \n",
    "  geom_point(size=7) +\n",
    "  geom_abline(slope=1,intercept=0,colour='black',linetype='dashed') +\n",
    "  theme_pubr() +\n",
    "  theme(axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=16),\n",
    "        legend.position = 'top',\n",
    "        legend.text = element_text(size=14),\n",
    "        legend.title = element_blank()) +\n",
    "  scale_colour_manual(values=b_colour) +\n",
    "  xlim(0,1) +\n",
    "  ylim(0,1) + \n",
    "  xlab('Global missense constraint') +\n",
    "  ylab('Regional missense constraint at hotspots')\n",
    "\n",
    "#Figure 5\n",
    "plot_grid(a,b,labels='AUTO',nrow=2,ncol=1,rel_widths = c(2,1),label_size = 14)\n",
    "\n",
    "#Put a p-value on regional missense constraint by comparing hotspot domains to ion channel domains with no statistical evidence of being a mutation hotspot\n",
    "mpc <- mpc_in[,mpc_starts:=ifelse(genomic_start<genomic_end,genomic_start,genomic_end)]\n",
    "mpc <- mpc[,mpc_ends:=ifelse(genomic_start<genomic_end,genomic_end,genomic_start)]\n",
    "mpc <- mpc[,obs_exp:=as.numeric(sub(',','.',obs_exp))]\n",
    "mpc <- mpc[,gene_name:=gene]\n",
    "mpc <- mpc[,gencode_transcript:=transcript]\n",
    "mpc <- mpc[,c('gene_name','gencode_transcript','mpc_starts','mpc_ends','obs_exp','region_name')]\n",
    "\n",
    "all_ion_channel_domains <- fread('/Users/jh/Downloads/ion_channels.csv')\n",
    "all_ion_channel_domains <- all_ion_channel_domains[,c('gene_name','chromosome','chromosome_position','codon_base_pair_position','gencode_transcription_id','strand')]\n",
    "\n",
    "getOverlaps <- function( gene, transcript ) {\n",
    "  \n",
    "  mpc_temp <- mpc[gene_name==gene & gencode_transcript==transcript]\n",
    "  \n",
    "  temp_ion_channel_domains <- all_ion_channel_domains[gene_name==gene & gencode_transcription_id==transcript]\n",
    "  strand =  unique(temp_ion_channel_domains[,strand])\n",
    "\n",
    "  if( strand=='minus' ) {\n",
    "    \n",
    "      temp_ion_channel_domains_sorted <- temp_ion_channel_domains[order(chromosome_position,decreasing = TRUE)]\n",
    "    \n",
    "  } else {\n",
    "    \n",
    "      temp_ion_channel_domains_sorted <- temp_ion_channel_domains[order(chromosome_position,decreasing = FALSE)]\n",
    "    \n",
    "  }\n",
    "  \n",
    "  assignment_vec <- rep(c(rep('start',3),rep('stop',3)),nrow(temp_ion_channel_domains)/6)\n",
    "  temp_ion_channel_domains_sorted <- temp_ion_channel_domains_sorted[,assignment:=assignment_vec]\n",
    "  \n",
    "  if( strand=='minus' ) {\n",
    "    \n",
    "    end <- temp_ion_channel_domains_sorted[gene_name==gene & gencode_transcription_id==transcript & assignment=='start' & codon_base_pair_position==0,chromosome_position]\n",
    "    start <- temp_ion_channel_domains_sorted[gene_name==gene & gencode_transcription_id==transcript & assignment=='stop'& codon_base_pair_position==2,chromosome_position]\n",
    "    \n",
    "  } else {\n",
    "    \n",
    "    start <- temp_ion_channel_domains_sorted[gene_name==gene & gencode_transcription_id==transcript & assignment=='start' & codon_base_pair_position==0,chromosome_position]\n",
    "    end <- temp_ion_channel_domains_sorted[gene_name==gene & gencode_transcription_id==transcript & assignment=='stop'& codon_base_pair_position==2,chromosome_position]\n",
    "    \n",
    "  }\n",
    "  \n",
    "  domains_temp <- data.table(gene = gene, transcript = transcript, domain_starts = start, domain_ends = end)\n",
    "  \n",
    "  setkey(mpc_temp,mpc_starts,mpc_ends)\n",
    "  setkey(domains_temp,domain_starts,domain_ends)\n",
    "  \n",
    "  overlaps <- foverlaps(domains_temp, mpc_temp, nomatch = 0)\n",
    "  \n",
    "  overlaps <- overlaps[,c('gene_name','gencode_transcript','obs_exp','region_name','mpc_starts','mpc_ends','domain_starts','domain_ends')]\n",
    "  return(overlaps)\n",
    "  \n",
    "}\n",
    "\n",
    "genes <- unique(all_ion_channel_domains[,gene_name])\n",
    "overlaps_out <- vector('list',length=nrow(all_ion_channel_domains))\n",
    "n = 1\n",
    "\n",
    "for( gene in genes ) {\n",
    "  \n",
    "  transcripts <- unique(all_ion_channel_domains[gene_name==gene,gencode_transcription_id])\n",
    "  \n",
    "  for( transcript in transcripts ) {\n",
    "    \n",
    "    overlaps_out[[n]] <- getOverlaps(gene, transcript)\n",
    "    n = n + 1\n",
    "    \n",
    "  }\n",
    "  \n",
    "}\n",
    "\n",
    "overlaps <- do.call(rbind,overlaps_out)\n",
    "\n",
    "overlaps <- overlaps[,class:=ifelse(gene_name %in% hotspot_genes,'Hotspot Genes',\n",
    "                                    ifelse(!gene_name %in% c(hotspot_genes, novel_hotspot_genes, dd_genes),'Control Genes','Other'))]\n",
    "\n",
    "overlaps_to_test <- overlaps[class=='Hotspot Genes' | class=='Control Genes']\n",
    "\n",
    "overlaps_labels <- c('PF00520 Domain-Containing\\nControl Genes','Hotspot Genes')\n",
    "\n",
    "ggplot(overlaps_to_test,aes(x=class,y=obs_exp,fill=class)) + \n",
    "  geom_boxplot(alpha=0.8) +\n",
    "  theme_pubr() +\n",
    "  theme(legend.position = 'none',\n",
    "        axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=14), \n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  ggsignif::geom_signif(\n",
    "    comparisons = list(c('Hotspot Genes','Control Genes')),\n",
    "    y_position = 1.3,\n",
    "    colour='black',\n",
    "    test = 'wilcox.test') +\n",
    "  scale_x_discrete(labels=overlaps_labels) +\n",
    "  scale_fill_manual(values=c('skyblue3','darkgoldenrod')) +\n",
    "  ylab('Observed/expected regional missense constraint at PF00520 domains') +\n",
    "  xlab('')\n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Figure 5. Proposed novel hotspot genes are constrained against loss of function and missense variation.** *A) Hotspot genes and proposed novel hotspot genes show population evidence of constraint.* The observed variant counts for loss of function (red), missense (orange), and synonymous (pink) variants from the gnomAD v2 release were compared to the expected counts based on a null mutational model (Samocha et. al., 2017). Points represent the mean observed/expected ratio for all genes in each set and bars denote the mean upper and lower bound fractions for these ratios. The dashed line at observed/expected = 1 indicates perfect adherence to the null mutational model (observed counts = expected counts); values that fall below this line are constrained. *B) Mutation hotspots occur in missense constrained regions within genes.* While the missense constraint values in Figure 5A are calculated across the entire coding region, some genes (n = 2700) show evidence of regional missense constraint (Samocha et. al., 2019). These genes can be statistically divided into regions with different observed/expected ratios for missense variation. 19 out of 25 hotspot and proposed novel hotspot genes show evidence of regional missense constraint (16 hotspot genes and 3 proposed novel hotspot genes). Points represent regions of differential missense constraint that contain mutation hotspots in these genes coloured by gene family. The global missense constraint score (observed/expected counts across the coding region, x-axis) is compared against the regional missense constraint score of these regions (observed/expected counts across differentially constrained region, y-axis). The dashed line at regional observed/expected = genic observed/expected indicates a regional missense constraint equal to the overall constraint score of the gene; points that fall below it have higher constraint against missense variation at mutation hotspots. \n",
    "\n",
    "### Supplementary figures\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Supplement A\n",
    "\n",
    "fig_a_supp <- gene_tpm %>%\n",
    "  filter(class!='Control Genes' | hgnc %in% sampled_genes | contains_ion_channel==TRUE) %>%\n",
    "  filter(contains_ion_channel==TRUE) %>%\n",
    "  filter(class!='Proposed Novel Hotspot Genes') %>%\n",
    "  group_by(tissue,class) %>%\n",
    "  summarise(\n",
    "    prop_expressed = sum(tpm > 1, na.rm=T) / n(),\n",
    "    .groups = 'keep'\n",
    "  ) %>%\n",
    "  data.table()\n",
    "\n",
    "#Figure 4A\n",
    "a_supp_colour <- c('Control Genes' = 'skyblue3',\n",
    "            'DD Consensus Genes' = 'darkgreen',\n",
    "            'Hotspot Genes' = 'darkgoldenrod')\n",
    "\n",
    "a_supp_labels <- c('Control Genes' = 'PF00520 Domain-Containing Control Genes',\n",
    "            'DD Consensus Genes' = 'PF00520 Domain-Containing NDD-Associated Genes',\n",
    "            'Hotspot Genes' = 'Hotspot Genes')\n",
    "\n",
    "a <- ggplot(fig_a_supp,aes(x=tissue,y=prop_expressed,colour=class)) +\n",
    "  geom_point(size=4) +\n",
    "  theme_pubr() +\n",
    "  theme(axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=16), \n",
    "        axis.text.x = element_text(angle = 90, \n",
    "                                   vjust = 0.5, \n",
    "                                   hjust=1),\n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_colour_manual(values=a_supp_colour,labels=a_supp_labels) +\n",
    "  ylim(0,1) +\n",
    "  xlab('') +\n",
    "  ylab('Proportion expressed genes (TPM > 1)')\n",
    "\n",
    "#Supplement N\n",
    "fig_b_supp_1 <- median_gene_tpm %>%\n",
    "  group_by(class) %>%\n",
    "  summarise(\n",
    "    prop_brain = sum(Brain > `Other Tissues`, na.rm=T) / n(),\n",
    "    prop_other = sum(`Other Tissues` > Brain, na.rm=T) / n(),\n",
    "    prop_unexpressed = sum(Brain == `Other Tissues`, na.rm=T) / n(),\n",
    "    .groups='drop_last'\n",
    "  ) %>%\n",
    "  reshape2::melt(id.vars=c('class')) %>%\n",
    "  data.table()\n",
    "\n",
    "#Supplement N\n",
    "fig_b_supp_2 <- median_gene_tpm %>%\n",
    "  filter(contains_ion_channel==TRUE) %>%\n",
    "  group_by(class) %>%\n",
    "  summarise(\n",
    "    num_brain = sum(Brain > `Other Tissues`, na.rm=T),\n",
    "    num_other_tissues = sum(`Other Tissues` > Brain, na.rm=T),\n",
    "    total = n(),\n",
    "    num_unexpressed = total - (num_brain + num_other_tissues),\n",
    "    prop_brain = sum(Brain > `Other Tissues`, na.rm=T) / n(),\n",
    "    prop_other = sum(`Other Tissues` > Brain, na.rm=T) / n(),\n",
    "    prop_unexpressed = sum(Brain == `Other Tissues`, na.rm=T) / n(),\n",
    "    .groups='drop_last'\n",
    "  ) %>%\n",
    "  reshape2::melt(id.vars=c('class')) %>%\n",
    "  data.table()\n",
    "\n",
    "#Fisher's exact: Hotspot genes versus ion channel containing control/DD consensus genes\n",
    "hotspot_counts_PF00520 <- c(fig_b_supp_2[class=='Hotspot Genes' & variable=='num_brain',value],\n",
    "                    fig_b_supp_2[class=='Hotspot Genes' & variable=='num_other_tissues',value])\n",
    "dd_counts_PF00520 <- c(fig_b_supp_2[class=='DD Consensus Genes' & variable=='num_brain',value],\n",
    "                       fig_b_supp_2[class=='DD Consensus Genes' & variable=='num_other_tissues',value])\n",
    "control_counts_PF00520 <- c(fig_b_supp_2[class=='Control Genes' & variable=='num_brain',value],\n",
    "                    fig_b_supp_2[class=='Control Genes' & variable=='num_other_tissues',value])\n",
    "\n",
    "fisher.test(cbind(hotspot_counts_PF00520,control_counts_PF00520))\n",
    "fisher.test(cbind(hotspot_counts_PF00520,dd_counts_PF00520))\n",
    "\n",
    "#Supplement N\n",
    "bsupp_x_labels = c('PF00520-Containing\\nControl Genes','PF00520-Containing\\nDD Consensus Genes','Hotspot Genes','Proposed Novel\\nHotspot Genes')\n",
    "bsupp_fill = c('prop_unexpressed' = 'darkgrey',\n",
    "         'prop_other' = 'steelblue3',\n",
    "         'prop_brain' = 'steelblue4')\n",
    "bsupp_levels = c('prop_unexpressed','prop_other','prop_brain')\n",
    "bsupp_fill_labels = c('prop_unexpressed' = 'Unexpressed',\n",
    "                'prop_other' = 'Other Tissues TPM > Brain TPM',\n",
    "                'prop_brain' = 'Brain TPM > Other Tissues TPM')\n",
    "\n",
    "ggplot(fig_b_supp_2[variable %in% c('prop_brain','prop_other','prop_unexpressed')],aes(x=class,y=value,fill=factor(variable,levels=bsupp_levels))) + \n",
    "  geom_col(colour='black',alpha=0.8) +\n",
    "  theme_pubr() +\n",
    "  theme(legend.position = 'top',\n",
    "        axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=14), \n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_fill_manual(labels=bsupp_fill_labels,values=bsupp_fill) +\n",
    "  scale_x_discrete(labels=bsupp_x_labels) +\n",
    "  xlab('') +\n",
    "  ylab('Proportion of Genes')\n",
    "\n",
    "#Supplmenetary table N\n",
    "median_gene_tpm_table <- median_gene_tpm %>%\n",
    "  filter(class=='Hotspot Genes' | class=='Proposed Novel Hotspot Genes') %>%\n",
    "  select(-c('contains_ion_channel')) %>%\n",
    "  plyr::arrange(class,wilcox_corrected_p) %>%\n",
    "  data.table()\n",
    "\n",
    "colnames(median_gene_tpm_table) <- c('Gene','Class','Wilcox P','Bonferroni-Corrected Wilcox P','Median TPM Brain','Median TPM Other Tissues')\n",
    "\n",
    "knitr::kable(median_gene_tpm_table)\n",
    "\n",
    "#Supplement N\n",
    "total_genes_rmc = length(unique(mpc_in[,gene]))\n",
    "dd_consensus_with_rmc = length(unique(mpc_in[gene %in% dd_genes,gene]))\n",
    "hotspot_with_rmc = length(unique(mpc_in[gene %in% hotspot_genes,gene]))\n",
    "novel_hotspot_with_rmc = length(unique(mpc_in[gene %in% novel_hotspot_genes,gene]))\n",
    "control_genes_with_rmc = total_genes_rmc - (dd_consensus_with_rmc + hotspot_with_rmc + novel_hotspot_with_rmc)\n",
    "\n",
    "hotspot_counts <- c(hotspot_with_rmc,length(hotspot_genes))\n",
    "dd_consensus_counts <- c(dd_consensus_with_rmc,length(dd_genes))\n",
    "control_counts = c(control_genes_with_rmc,length(unique(gene_tpm[class=='Control Genes',hgnc])))\n",
    "\n",
    "supp_c <- data.table(rmc_counts=c(novel_hotspot_with_rmc,hotspot_with_rmc,dd_consensus_with_rmc,control_genes_with_rmc),\n",
    "                     total_counts=c(length(novel_hotspot_genes),length(hotspot_genes),length(dd_genes),length(unique(gene_tpm[class=='Control Genes',hgnc]))),\n",
    "                     class=c('Proposed Novel Hotspot Genes','Hotspot Genes','DD Consensus Genes','Control Genes'))\n",
    "\n",
    "supp_c <- supp_c %>%\n",
    "  mutate(prop_rmc = rmc_counts/total_counts,\n",
    "         prop_other = 1 - prop_rmc) %>%\n",
    "  reshape2::melt(id.vars='class') %>%\n",
    "  data.table()\n",
    "\n",
    "supp_c_x_labels = c('Control Genes','DD Consensus Genes','Hotspot Genes','Proposed Novel\\nHotspot Genes')\n",
    "supp_c_labels = c('Other Genes','Proportion of Genes with Evidence of Regional Missense Constraint')\n",
    "supp_c_fill = c('darkgrey','tomato4')\n",
    "\n",
    "ggplot(supp_c[variable=='prop_rmc' | variable=='prop_other'],\n",
    "       aes(x=class,y=value,fill=factor(variable,levels=c('prop_other','prop_rmc')))) +\n",
    "  geom_col() +\n",
    "  theme_pubr() +\n",
    "  theme(legend.position = 'top',\n",
    "        axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=14), \n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_fill_manual(values=supp_c_fill,labels=supp_c_labels) +\n",
    "  scale_x_discrete(labels=supp_c_x_labels) +\n",
    "  xlab('') +\n",
    "  ylab('Proportion of Genes')\n",
    "\n",
    "fisher.test(cbind(hotspot_counts,control_counts))\n",
    "fisher.test(cbind(hotspot_counts,dd_consensus_counts))\n",
    "fisher.test(cbind(dd_consensus_counts,control_counts))\n",
    "\n",
    "#Enrichment of activating mutations analysis on DDG2P\n",
    "\n",
    "gene_list <- fread('/Users/jh/Resources/Gene_Lists/DDG2P_22-4-2021.csv')\n",
    "nonhap_list <- c('dominant negative','gain of function','activating','increased gene dosage')\n",
    "\n",
    "ddg2p_genes <- length(unique(gene_list[,`gene symbol`]))\n",
    "ddg2p_activating_genes <- length(unique(gene_list[`mutation consequence` %in% nonhap_list,`gene symbol`]))\n",
    "hotspot_ddg2p_genes <- length(unique(gene_list[`gene symbol` %in% hotspot_genes,`gene symbol`]))\n",
    "hotspot_ddg2p_activating_genes <- length(unique(gene_list[`gene symbol` %in% hotspot_genes & `mutation consequence` %in% nonhap_list,`gene symbol`]))\n",
    "\n",
    "fisher.test(cbind(c(ddg2p_activating_genes,(ddg2p_genes - ddg2p_activating_genes)),\n",
    "                  c(hotspot_ddg2p_activating_genes,(hotspot_ddg2p_genes - hotspot_ddg2p_activating_genes))))\n",
    "\n",
    "#NDD genes with ion channel domain in the lenient set -- for discussion\n",
    "\n",
    "test <- median_gene_tpm %>%\n",
    "  filter(class!='Control Genes' & class!='Proposed Novel Hotspot Genes' & contains_ion_channel==TRUE) %>%\n",
    "  mutate(class = ifelse(class=='DD Consensus Genes' & contains_ion_channel==TRUE,'NDD-Associated PF00520 Domain-Containing Genes',class)) %>%\n",
    "  reshape2::melt(measure.vars=c('Brain','Other Tissues')) %>%\n",
    "  data.table()\n",
    "\n",
    "ggplot(test,aes(x=class,y=value,fill=variable)) + \n",
    "  geom_boxplot() +\n",
    "  theme_pubr()\n",
    "\n",
    "#Test of highest TPM tissue\n",
    "\n",
    "max_tpm_tissue <- gene_tpm %>%\n",
    "  group_by(hgnc) %>%\n",
    "  filter(tpm==max(tpm)) %>%\n",
    "  mutate(tissue_type = ifelse(tissue %in% brain_tissues,'Brain','Other Tissues')) %>%\n",
    "  ungroup() %>%\n",
    "  group_by(class) %>%\n",
    "  summarise(\n",
    "    prop_brain = sum(tissue_type=='Brain', na.rm=T) / n(),\n",
    "    prop_other = sum(tissue_type=='Other Tissues', na.rm=T) / n(),\n",
    "    .groups = 'keep'\n",
    "  ) %>%\n",
    "  data.table()\n",
    "\n",
    "brain_expressed_genes <- gene_tpm %>%\n",
    "  group_by(ensembl_versioned) %>%\n",
    "  filter(tissue %in% brain_tissues & tpm > 1) %>%\n",
    "  ungroup() %>%\n",
    "  pull(hgnc)\n",
    "  \n",
    "num_other_expressed_tissues <- gene_tpm %>%\n",
    "  mutate(class = ifelse(contains_ion_channel==TRUE,paste0(class,' - TRUE'),class)) %>%\n",
    "  filter(tpm > 1) %>%\n",
    "  group_by(ensembl_versioned) %>%\n",
    "  mutate(num_expressed = n(),\n",
    "         prop_expressed = num_expressed/41) %>%\n",
    "  data.table()\n",
    "\n",
    "num_other_expressed_tissues_brain <- gene_tpm %>%\n",
    "  filter(hgnc %in% brain_expressed_genes) %>%\n",
    "  filter(!tissue %in% brain_tissues) %>%\n",
    "  mutate(class = ifelse(contains_ion_channel==TRUE,paste0(class,' - TRUE'),class)) %>%\n",
    "  filter(tpm > 1) %>%\n",
    "  group_by(ensembl_versioned) %>%\n",
    "  mutate(num_expressed = n(),\n",
    "         prop_expressed = num_expressed/41) %>%\n",
    "  data.table()\n",
    "  \n",
    "labels_1 = c('Control Genes' = 'Control Genes\\nExpressed',\n",
    "           'Control Genes - TRUE' = 'PF00520 Domain-Containing\\nControl Genes Expressed',\n",
    "           'DD Consensus Genes' = 'NDD-Associated Genes\\nExpressed',\n",
    "           'DD Consensus Genes - TRUE' = 'PF00520 Domain-Containing\\nNDD-Associated Genes\\nExpressed',\n",
    "           'Hotspot Genes - TRUE' = 'Hotspot Genes\\nExpressed',\n",
    "           'Proposed Novel Hotspot Genes - TRUE' = 'Proposed Novel\\nHotspot Genes Expressed')\n",
    "\n",
    "labels_2 = c('Control Genes' = 'Control Genes\\nExpressed in Brain',\n",
    "           'Control Genes - TRUE' = 'PF00520 Domain-Containing\\nControl Genes Expressed in Brain',\n",
    "           'DD Consensus Genes' = 'NDD-Associated Genes\\nExpressed in Brain',\n",
    "           'DD Consensus Genes - TRUE' = 'PF00520 Domain-Containing\\nNDD-Associated Genes\\nExpressed in Brain',\n",
    "           'Hotspot Genes - TRUE' = 'Hotspot Genes\\nExpressed in Brain',\n",
    "           'Proposed Novel Hotspot Genes - TRUE' = 'Proposed Novel\\nHotspot Genes Expressed in Brain')\n",
    "\n",
    "supp_a <- ggplot(num_other_expressed_tissues,aes(x=class,y=num_expressed,fill=class)) +\n",
    "  geom_boxplot(alpha=0.8) + \n",
    "  theme_pubr() +\n",
    "  theme(legend.position = 'none',\n",
    "        axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=14), \n",
    "        legend.title = element_blank(),\n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_x_discrete(labels=labels_1) +\n",
    "  scale_fill_manual(values=c('skyblue3','skyblue3','darkgreen','darkgreen','darkgoldenrod','darkgoldenrod'),labels=labels_1) +\n",
    "  ggsignif::geom_signif(\n",
    "    comparisons = list(c('Control Genes','Control Genes - TRUE'),c('DD Consensus Genes','DD Consensus Genes - TRUE'),\n",
    "                       c('DD Consensus Genes - TRUE','Hotspot Genes - TRUE')),\n",
    "    map_signif_level = FALSE,\n",
    "    y_position = c(45,45,43),\n",
    "    colour='black') +\n",
    "  xlab('') +\n",
    "  ylab('Number of Tissues with Expression (TPM > 1)')\n",
    "\n",
    "supp_b <- ggplot(num_other_expressed_tissues_brain,aes(x=class,y=num_expressed,fill=class)) +\n",
    "  geom_boxplot(alpha=0.8) + \n",
    "  theme_pubr() +\n",
    "  theme(legend.position = 'none',\n",
    "        axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=14), \n",
    "        legend.title = element_blank(),\n",
    "        legend.text = element_text(size=14)) +\n",
    "  scale_x_discrete(labels=labels_2) +\n",
    "  scale_fill_manual(values=c('skyblue3','skyblue3','darkgreen','darkgreen','darkgoldenrod','darkgoldenrod'),labels=labels_2) +\n",
    "  ggsignif::geom_signif(\n",
    "    comparisons = list(c('Control Genes','Control Genes - TRUE'),c('DD Consensus Genes','DD Consensus Genes - TRUE'),\n",
    "                       c('DD Consensus Genes - TRUE','Hotspot Genes - TRUE')),\n",
    "    map_signif_level = FALSE,\n",
    "    y_position = c(45,45,43),\n",
    "    colour='black') +\n",
    "  xlab('') +\n",
    "  ylab('Number of Non-Brain Tissues with Expression (TPM > 1)')\n",
    "\n",
    "plot_grid(supp_a,supp_b,nrow=2,labels='AUTO')\n",
    "\n",
    "\n",
    "skeletal_muscle_genes <- c('SCN4A','RYR1','CACNA1S','KCNA7')\n",
    "\n",
    "test <- gene_tpm %>%\n",
    "  filter(hgnc %in% skeletal_muscle_genes) %>%\n",
    "  filter(tpm > 1) %>%\n",
    "  data.table()\n",
    "\n",
    "ggplot(test,aes(x=tissue,y=tpm,colour=hgnc)) + \n",
    "  geom_point(size=5,alpha=0.6) +\n",
    "  geom_hline(yintercept=1,linetype='dashed',colour='black') +\n",
    "  theme_pubr() +\n",
    "  theme(axis.text = element_text(size=14), \n",
    "        axis.title = element_text(size=16), \n",
    "        axis.text.x = element_text(angle = 90, \n",
    "                                   vjust = 0.5, \n",
    "                                   hjust=1),\n",
    "        legend.title = element_blank(), \n",
    "        legend.text = element_text(size=14)) +\n",
    "  xlab('') +\n",
    "  ylab('TPM of Expressed Genes (TPM > 1)')\n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": "",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
