{
    "cells": [
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": "\n"
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "knitr::opts_chunk$set(echo = TRUE)\n",
                "require(tidyverse)\n",
                "require(data.table)\n",
                "require(reshape2)\n",
                "\n",
                "#Install additional packages\n",
                "install.packages(\"ggpubr\")\n",
                "install.packages(\"here\")\n",
                "install.packages(\"cowplot\")\n",
                "install.packages(\"R.utils\")\n",
                "require(ggpubr)\n",
                "require(here)\n",
                "require(cowplot)\n",
                "require(R.utils)\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": "\n"
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Project generated\n",
                "knitr::opts_chunk$set(echo = TRUE)\n",
                "dd_genes <- fread(here('data/discordant_consensus_novel_DD_associated_genes_kaplanis2020.txt'),header=FALSE,data.table=TRUE)[,V1]\n",
                "hotspot_genes <- fread(here('data/hotspot_genes.txt'),header=FALSE,data.table=TRUE)[,V1]\n",
                "novel_hotspot_genes <- fread(here('data/proposed_novel_hotspot_genes.txt'),header=FALSE,data.table=TRUE)[,V1]\n",
                "ion_channel_genes <- fread(here('data/All_genes_with_PF00520_domain.txt'),header=FALSE,data.table=TRUE)[,V1]\n",
                "hotspot_genetic_positions <- fread(here('data/hotspot_to_genomic_positions_PF00520.txt',header=TRUE,data.table=TRUE))\n",
                "\n",
                "#External datasets\n",
                "gene_tpm <- fread(here('data/external/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz'),header=TRUE,data.table=TRUE)\n",
                "constraint_in <- fread(here('data/external/gnomad.v2.1.1.lof_metrics.by_gene.txt'),header=TRUE,data.table=TRUE)\n",
                "mpc_in <- fread(here('data/external/mpc_missense_constraint_multiple_regions_samocha17.txt'),header=TRUE,data.table=TRUE)\n",
                "ddg2p_in <- fread(here('data/external/DDG2P_22-4-2021.csv'),header=TRUE,data.table=TRUE)\n",
                "omim_in <- fread(here('data/external/genemap2.txt'),header=TRUE,data.table=TRUE)\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "### Expression input data and counts\n",
                "\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Remove novel hotspot genes from hotspot gene list\n",
                "hotspot_genes <- hotspot_genes[!hotspot_genes %in% novel_hotspot_genes]\n",
                "\n",
                "#Remove hotspot genes from DD consensus gene list\n",
                "dd_genes <- dd_genes[!dd_genes %in% hotspot_genes]\n",
                "\n",
                "#Parse gene TPM file\n",
                "gene_tpm <- gene_tpm[,Gene_ID:=gsub(\"\\\\..*\",\"\",Name)]\n",
                "gene_tpm <- data.table(melt(gene_tpm,id.vars = c('Description','Name','Gene_ID')))\n",
                "colnames(gene_tpm) <- c('hgnc','ensembl_versioned','ensembl','tissue','tpm')\n",
                "\n",
                "gene_tpm <- gene_tpm[,class:=ifelse(hgnc %in% novel_hotspot_genes,'Proposed Novel Hotspot Genes',\n",
                "                                    ifelse(hgnc %in% hotspot_genes,'Hotspot Genes',\n",
                "                                           ifelse(hgnc %in% dd_genes,'NDD-Associated Genes','Control Genes')))]\n",
                "\n",
                "gene_tpm <- gene_tpm[,contains_ion_channel:=ifelse(hgnc %in% ion_channel_genes,TRUE,FALSE)]\n",
                "\n",
                "head(gene_tpm)\n",
                "\n",
                "gene_tpm_tab <- gene_tpm %>%\n",
                "  group_by(class) %>%\n",
                "  dplyr::select('hgnc','class') %>%\n",
                "  unique() %>%\n",
                "  count()\n",
                "\n",
                "colnames(gene_tpm_tab) <- c('Class','Count')\n",
                "  \n",
                "#Table of counts by class (Control Genes, DD Consensus Genes, Hotspot Genes, Proposed Novel Hotspot Genes)\n",
                "knitr::kable(gene_tpm_tab)\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "### Figure 5A: Tissue expression of hotspot genes compared to control and other NDD genes\n",
                "\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Set seed\n",
                "set.seed(1255)\n",
                "\n",
                "#Randomly sample NDD-associated and control Genes\n",
                "sampleGenes <- function( iterations, sampling_size ) {\n",
                "\n",
                "  out <- vector('list',length=iterations)\n",
                "  \n",
                "  for( i in 1:iterations ) {\n",
                "    \n",
                "    sampled_dat <- gene_tpm %>% \n",
                "    filter(class!='Hotspot Genes' & class!='Proposed Novel Hotspot Genes') %>%\n",
                "    group_by(tissue,class) %>% \n",
                "    sample_n(sampling_size,replace=TRUE) %>%\n",
                "    mutate(i=factor(i))\n",
                "    \n",
                "    out[[i]] <- data.table(sampled_dat)\n",
                "    \n",
                "  }\n",
                "  \n",
                "  out_dt <- do.call(rbind,out)\n",
                "  return(out_dt)\n",
                "\n",
                "}\n",
                "\n",
                "hotspot_tpm <- gene_tpm %>%\n",
                "  filter(class=='Hotspot Genes') %>%\n",
                "  mutate(i=0)\n",
                "\n",
                "#Bind sampled genes to hotspot genes; calculate the proportion expressed, mean, and standard deviation across sets\n",
                "fig_5a <- rbind(sampleGenes(1000,19),hotspot_tpm) %>%\n",
                "  group_by(tissue,class,i) %>%\n",
                "  summarise(\n",
                "    prop_expressed = sum(tpm > 1, na.rm=T) / n(),\n",
                "    .groups = 'drop_last'\n",
                "  ) %>%\n",
                "  ungroup() %>%\n",
                "  group_by(tissue,class) %>%\n",
                "  summarise(\n",
                "    mean = mean(prop_expressed),\n",
                "    min = mean - sd(prop_expressed),\n",
                "    max = mean + sd(prop_expressed),\n",
                "    .groups = 'keep'\n",
                "  ) %>%\n",
                "  data.table()\n",
                "\n",
                "#Figure 5A\n",
                "a_fill <- c('Control Genes' = 'skyblue3',\n",
                "            'NDD-Associated Genes' = 'darkgreen',\n",
                "            'Hotspot Genes' = 'darkgoldenrod')\n",
                "\n",
                "a <- ggplot(fig_5a[class!='Proposed Novel Hotspot Genes' & class!='Hotspot Genes'],aes(x=tissue,y=mean,colour=class,ymin=min,ymax=max)) +\n",
                "  geom_point(alpha=0.7,size=3,shape=15) +\n",
                "  geom_linerange(alpha=0.5,size=3) + \n",
                "  geom_point(data=fig_a[class=='Hotspot Genes'],aes(x=tissue,y=mean,colour=class),size=5) +\n",
                "  theme_pubr() +\n",
                "  theme(axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=16), \n",
                "        axis.text.x = element_text(angle = 90, \n",
                "                                   vjust = 0.5, \n",
                "                                   hjust=1),\n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  scale_colour_manual(values=a_fill,\n",
                "                      guide=guide_legend(override.aes = list(linetype = c(1,0,1),\n",
                "                                                           shape = c(15,19,15),\n",
                "                                                           size = c(3,5,3)))) +\n",
                "  ylim(0,1) +\n",
                "  xlab('') +\n",
                "  ylab('Proportion expressed genes (TPM > 1)')\n",
                "\n",
                "plot_grid(a,labels='AUTO',label_size=16)\n",
                "ggsave('~/results/main/Figure_5A.svg',plot=last_plot(),device='svg',dpi=300,width=14,height=10)\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "### Figure 5B: TPM distribution in brain and other tissues varies across gene sets\n",
                "\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Compute median TPM in brain and other tissues\n",
                "median_gene_tpm <- gene_tpm %>%\n",
                "  mutate(tissue_group = ifelse(tissue %in% brain_tissues,'Brain','Other Tissues')) %>%\n",
                "  group_by(ensembl_versioned,tissue_group) %>%\n",
                "  mutate(median_tpm_tissuegroup = median(tpm)) %>%\n",
                "  dplyr::select(-c('tpm','tissue')) %>%\n",
                "  unique() %>%\n",
                "  reshape2::dcast(formula=hgnc + class + ensembl_versioned + contains_ion_channel ~ tissue_group, value.var='median_tpm_tissuegroup') %>%\n",
                "  mutate(class = ifelse(hgnc %in% ion_channel_genes & class=='NDD-Associated Genes','PF00520-Containing NDD-Associated Genes',\n",
                "                        ifelse(hgnc %in% ion_channel_genes & class=='Control Genes','PF00520-Containing Control Genes',class))) %>%\n",
                "  data.table()\n",
                "\n",
                "#Figure 5B\n",
                "labels_b = c('NDD-Associated Genes' = 'NDD-Associated Genes',\n",
                "           'PF00520-Containing NDD-Associated Genes' = 'PF00520 Domain-Containing\\nNDD-Associated Genes',\n",
                "           'Hotspot Genes' = 'Hotspot Genes',\n",
                "           'Proposed Novel Hotspot Genes' = 'Proposed Novel Hotspot Genes')\n",
                "\n",
                "b <- ggplot(median_gene_tpm,aes(x=log10(`Other Tissues`),\n",
                "                                y=log10(Brain),\n",
                "                                colour=factor(class,levels=c('Control Genes',\n",
                "                                                             'NDD-Associated Genes',\n",
                "                                                             'PF00520-Containing NDD-Associated Genes',\n",
                "                                                             'Hotspot Genes',\n",
                "                                                             'Proposed Novel Hotspot Genes')))) +\n",
                "  geom_abline(slope=1,intercept=0,linetype='dashed',colour='black') +\n",
                "  geom_density2d(data=median_gene_tpm[class=='Control Genes' & Brain!=0 & `Other Tissues`!=0],alpha=0.7) +\n",
                "  geom_density2d(data=median_gene_tpm[class=='NDD-Associated Genes' & Brain!=0 & `Other Tissues`!=0],alpha=0.7) +\n",
                "  geom_point(data=median_gene_tpm[class=='PF00520-Containing NDD-Associated Genes'],size=4) +\n",
                "  geom_point(data=median_gene_tpm[class=='Hotspot Genes'],size=4) +\n",
                "  geom_text(data=median_gene_tpm[class=='Proposed Novel Hotspot Genes'],\n",
                "            aes(label=hgnc),size=5,fontface='italic',show.legend=FALSE) +\n",
                "  theme_pubr() +\n",
                "  theme(axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=16), \n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14),\n",
                "        legend.position = c(0.75,0.13)) +\n",
                "  scale_colour_manual(labels=labels_b,values=c('skyblue3',\n",
                "                                               'darkgoldenrod',\n",
                "                                               'darkgreen',\n",
                "                                               'grey40',\n",
                "                                               'tomato4'),\n",
                "                      guide=guide_legend(override.aes = list(linetype=c(1,0,1,0,0),\n",
                "                                                             shape=c(NA,19,NA,19,65)),\n",
                "                                         nrow=5,\n",
                "                                         by_row=TRUE)) +\n",
                "  ylab('Log10 Median Expression in Brain Tissues') + \n",
                "  xlab('Log10 Median Expression in Other Tissues') +\n",
                "  coord_cartesian(xlim=c(-3,3),ylim=c(-3,3))\n",
                "\n",
                "plot_grid(b,labels='B',label_size=16)\n",
                "\n",
                "ggsave('~/results/main/Figure_5B.svg',plot=last_plot(),device='svg',dpi=300,width=14,height=10)\n",
                "\n",
                "plot_grid(a,b,labels='AUTO',label_size=16,nrow=1,rel_widths = c(1.7,1),align='hv',axis='t')\n",
                "\n",
                "ggsave('~/results/main/Figure_5.png',plot=last_plot(),device='png',dpi=300,width=20,height=9)\n",
                "ggsave('~/results/main/Figure_5.svg',plot=last_plot(),device='svg',dpi=300,width=20,height=9)\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "### Figure 4A: Constraint in hotspot and proposed novel hotspot genes\n",
                "\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Get relevant constraint information\n",
                "probabilistic_constraint <- c('pLI','pRec','pNull')\n",
                "upper_bound_fraction <- c('oe_lof_upper','oe_mis_upper','oe_syn_upper')\n",
                "observed_expected <- c('oe_lof','oe_mis','oe_syn','oe_lof_lower','oe_mis_lower','oe_syn_lower',upper_bound_fraction)\n",
                "z_scores <- c('lof_z','mis_z','syn_z')\n",
                "all_constraint <- unique(c(probabilistic_constraint,upper_bound_fraction,observed_expected,z_scores))\n",
                "\n",
                "constraint <- constraint_in %>%\n",
                "  subset(select=c('gene',all_constraint)) %>%\n",
                "  reshape2::melt(id.vars = 'gene', value.name = 'value', var.name = 'constraint_metric') %>%\n",
                "  filter(!is.na(value)) %>%\n",
                "  mutate(class=ifelse(gene %in% novel_hotspot_genes,'Proposed Novel Hotspot Genes',\n",
                "                      ifelse(gene %in% hotspot_genes,'Hotspot Genes',\n",
                "                             ifelse(gene %in% dd_genes,'NDD-Associated Genes','Control Genes'))))\n",
                "\n",
                "fig_4a <- constraint %>%\n",
                "  filter(variable %in% observed_expected) %>%\n",
                "  group_by(class,variable) %>%\n",
                "  mutate(variable=gsub('oe_','',variable)) %>%\n",
                "  mutate(type=substr(variable,1,3)) %>%\n",
                "  mutate(bound=ifelse(grepl('upper',variable),'upper',\n",
                "                      ifelse(grepl('lower',variable),'lower',''))) %>%\n",
                "  group_by(class,type,bound) %>%\n",
                "  summarise(\n",
                "    mean = mean(value),\n",
                "    .groups = 'drop_last'\n",
                "  ) %>%\n",
                "  mutate(bound=ifelse(bound=='','oe',bound)) %>%\n",
                "  reshape2::dcast(formula=class + type ~ bound, value.var='mean')\n",
                "\n",
                "#Figure 4A\n",
                "a_labels = c('Loss of function','Missense','Synonymous')\n",
                "a_levels = c('Control Genes','NDD-Associated Genes','Hotspot Genes','Proposed Novel Hotspot Genes')\n",
                "a_x_labels = c('Control Genes\\nn = 18648','NDD-Associated\\nGenes\\nn = 985','Hotspot Genes\\nn = 19','Proposed Novel\\nHotspot Genes\\nn = 6')\n",
                "\n",
                "a <- ggplot(fig_4a,aes(x=factor(class,levels=a_levels),y=oe,colour=type,ymin=lower,ymax=upper)) + \n",
                "  geom_pointrange(position=position_dodge(width=1),size=1.6) + \n",
                "  geom_hline(yintercept=1,linetype='dashed',colour='black') +\n",
                "  theme_pubr() +\n",
                "  theme(axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=16), \n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  scale_x_discrete(labels=a_x_labels) +\n",
                "  scale_colour_manual(labels=a_labels, values=c('tomato4','darkorange3','deeppink4')) + \n",
                "  ylab('Observed/expected counts') +\n",
                "  xlab('')\n",
                "\n",
                "ggsave('~/results/main/Figure_5A.svg',plot=last_plot(),device='svg',dpi=300,width=10,height=8)\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "### Figure 4B: Mutation hotspots occur in missense constrained regions within genes\n",
                "\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Get regional missense constraint at hotspot positions\n",
                "mpc <- mpc_in[,mpc_starts:=ifelse(genomic_start<genomic_end,genomic_start,genomic_end)]\n",
                "mpc <- mpc[,mpc_ends:=ifelse(genomic_start<genomic_end,genomic_end,genomic_start)]\n",
                "mpc <- mpc[,obs_exp:=as.numeric(sub(',','.',obs_exp))]\n",
                "mpc <- mpc[,gene_name:=gene]\n",
                "mpc <- mpc[,gencode_transcript:=transcript]\n",
                "mpc <- mpc[,c('gene_name','gencode_transcript','mpc_starts','mpc_ends','obs_exp','region_name')]\n",
                "\n",
                "getRegionalConstraint <- function( gene, transcript ) {\n",
                "  \n",
                "  mpc_temp <- mpc[gene_name==gene & gencode_transcript==transcript]\n",
                "  temp_hotspots <- hotspot_genetic_positions[hgnc==gene & gencode_transcription_id==transcript]\n",
                "  temp_hotspots <- temp_hotspots[,start_pos:=min(chromosome_position_base_pair_one,chromosome_position_base_pair_two,chromosome_position_base_pair_three),by=1:nrow(temp_hotspots)]\n",
                "  temp_hotspots <- temp_hotspots[,end_pos:=max(chromosome_position_base_pair_one,chromosome_position_base_pair_two,chromosome_position_base_pair_three),by=1:nrow(temp_hotspots)]\n",
                "\n",
                "  setkey(mpc_temp,mpc_starts,mpc_ends)\n",
                "  setkey(temp_hotspots,start_pos,end_pos)\n",
                "  \n",
                "  overlaps <- foverlaps(temp_hotspots, mpc_temp, nomatch = 0)\n",
                "  \n",
                "  overlaps <- overlaps[,c('gene_name','gencode_transcript','obs_exp','region_name','mpc_starts','mpc_ends','start_pos','end_pos','hotspot')]\n",
                "  return(overlaps)\n",
                "  \n",
                "}\n",
                "\n",
                "genes <- unique(hotspot_genetic_positions[,hgnc])\n",
                "regionalconstraint_out <- vector('list',length=nrow(hotspot_genetic_positions))\n",
                "n = 1\n",
                "\n",
                "for( gene in genes ) {\n",
                "  \n",
                "  transcripts <- unique(hotspot_genetic_positions[hgnc==gene,gencode_transcription_id])\n",
                "  \n",
                "  for( transcript in transcripts ) {\n",
                "    \n",
                "    regionalconstraint_out[[n]] <- getRegionalConstraint(gene, transcript)\n",
                "    n = n + 1\n",
                "    \n",
                "  }\n",
                "  \n",
                "}\n",
                "\n",
                "regional_constraint <- do.call(rbind,regionalconstraint_out)\n",
                "\n",
                "regional_constraint <- regional_constraint[,class:=ifelse(gene_name %in% hotspot_genes,'Hotspot Genes',\n",
                "                                    ifelse(gene_name %in% novel_hotspot_genes,'Proposed Novel Hotspot Genes',\n",
                "                                           ifelse(gene_name %in% dd_genes,'NDD-Associated Genes','Control Genes')))]\n",
                "\n",
                "colnames(regional_constraint) <- c('gene','gencode_transcript','obs_exp','region_name','mpc_starts','mpc_ends','hotspot_starts','hotspot_ends','hotspot','class')\n",
                "\n",
                "#Get global missense constraint\n",
                "global_oe_mis <- constraint %>%\n",
                "  filter(variable=='oe_mis')\n",
                "\n",
                "#Add global missense constraint to regional constraint infile and parse\n",
                "fig_4b <- regional_constraint %>%\n",
                "  inner_join(global_oe_mis,by='gene') %>%\n",
                "  dplyr::select(obs_exp,value,gene,class.x,hotspot) %>%\n",
                "  mutate(obs_exp=ifelse(obs_exp>1,1,obs_exp),\n",
                "         value=ifelse(value>1,1,value)) %>%\n",
                "  unique() %>%\n",
                "  data.table()\n",
                "\n",
                "#Figure 4B\n",
                "b_labels <- c('Control Genes' = 'PF00520 Domain-Containing\\nControl Genes\\nn = 28',\n",
                "              'Hotspot Genes' = 'Hotspot Genes\\nn = 16',\n",
                "              'NDD-Associated Genes' = 'PF00520 Domain-Containing\\nNDD-Associated Genes\\nn = 7')\n",
                "\n",
                "b <- ggplot(fig_4b[class.x!='Proposed Novel Hotspot Genes'],aes(x=class.x,y=obs_exp,fill=class.x)) +\n",
                "  geom_boxplot(alpha=0.8) +\n",
                "  theme_pubr() +\n",
                "  theme(legend.position = 'none',\n",
                "        axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=16), \n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  ggsignif::geom_signif(\n",
                "    comparisons = list(c('NDD-Associated Genes','Hotspot Genes'),\n",
                "                         c('Control Genes','Hotspot Genes')),\n",
                "    y_position = c(1.2,1.3),\n",
                "    colour='black',\n",
                "    test = 'wilcox.test') +\n",
                "  scale_x_discrete(labels=b_labels) +\n",
                "  scale_fill_manual(values=c('skyblue3','darkgoldenrod','darkgreen')) +\n",
                "  ylab('Observed/expected regional missense constraint\\nat PF00520 domains') +\n",
                "  xlab('')\n",
                "\n",
                "plot_grid(b,labels='B',label_size=16)\n",
                "\n",
                "ggsave('~/results/main/Figure_4B.svg',plot=last_plot(),device='svg',dpi=300,width=10,height=8)\n",
                "\n",
                "plot_grid(a,b,labels='AUTO',label_size=16,nrow=1,rel_widths = c(1.1,1),align='hv',axis='bt')\n",
                "\n",
                "ggsave('~/results/main/Figure_4.svg',plot=last_plot(),device='svg',dpi=300,width=18,height=8)\n",
                "ggsave('~/results/main/Figure_4.png',plot=last_plot(),device='png',dpi=300,width=18,height=8)\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "### Supplementary data\n",
                "\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Supplementary Data S9: Mutational constraint in hotspot and proposed novel hotspot genes\n",
                "supp_data_s9 <- constraint_in %>%\n",
                "  subset(select=c('gene',all_constraint)) %>%\n",
                "  drop_na() %>%\n",
                "  mutate(class=ifelse(gene %in% novel_hotspot_genes,'Proposed Novel Hotspot Genes',\n",
                "                      ifelse(gene %in% hotspot_genes,'Hotspot Genes',\n",
                "                             ifelse(gene %in% dd_genes,'NDD-Associated Genes','Control Genes')))) %>%\n",
                "  filter(class=='Hotspot Genes' | class=='Proposed Novel Hotspot Genes') %>%\n",
                "  arrange(class,pLI) %>%\n",
                "  data.table()\n",
                "\n",
                "write.table(supp_data_s9,'~/results/supp/Supplementary_Data_S9_Constraint.txt',col.names=TRUE,\n",
                "            row.names=FALSE,quote=FALSE,sep='\\t')\n",
                "\n",
                "#Supplementary Data S10: Proportion of hotspot genes expressed across tissues\n",
                "prop_expressed <- gene_tpm %>%\n",
                "  group_by(tissue,class) %>%\n",
                "  summarise(\n",
                "    num_expressed = sum(tpm > 1, na.rm=T),\n",
                "    num_unexpressed = sum(tpm <= 1, na.rm=T),\n",
                "    .groups = 'drop_last'\n",
                "  ) %>%\n",
                "  filter(class!='Proposed Novel Hotspot Genes') %>%\n",
                "  data.table()\n",
                "\n",
                "getFisher <- function( class1, class2, tissue_type ) {\n",
                "  \n",
                "  num_expressed_1 <- unique(prop_expressed[class==class1 & tissue==tissue_type,num_expressed])\n",
                "  num_expressed_2 <- unique(prop_expressed[class==class2 & tissue==tissue_type,num_expressed])\n",
                "  \n",
                "  num_unexpressed_1 <- unique(prop_expressed[class==class1 & tissue==tissue_type,num_unexpressed])\n",
                "  num_unexpressed_2 <- unique(prop_expressed[class==class2 & tissue==tissue_type,num_unexpressed])\n",
                "  \n",
                "  class1_prop <- c(num_expressed_1,num_unexpressed_1)\n",
                "  class2_prop <- c(num_expressed_2,num_unexpressed_2)\n",
                "  \n",
                "  p <- fisher.test(cbind(class1_prop,class2_prop))$p.value\n",
                "  return(p)\n",
                "  \n",
                "}\n",
                "\n",
                "prop_expressed <- prop_expressed[,Fisher_p_controls:=getFisher('Hotspot Genes','Control Genes',tissue),by=tissue]\n",
                "prop_expressed <- prop_expressed[,Fisher_p_NDD:=getFisher('Hotspot Genes','NDD-Associated Genes',tissue),by=tissue]\n",
                "prop_expressed <- prop_expressed[,Fisher_p_controls_bonferroni:=Fisher_p_controls*54] #Bonferroni correct over 54 GTEx tissues\n",
                "prop_expressed <- prop_expressed[,Fisher_p_NDD_bonferroni:=Fisher_p_NDD*54] #Bonferroni correct over 54 GTEx tissues\n",
                "\n",
                "supp_data_s10 <- prop_expressed %>%\n",
                "  mutate(Fisher_p_controls_bonferroni=ifelse(Fisher_p_controls_bonferroni>1,1,Fisher_p_controls_bonferroni),\n",
                "         Fisher_p_NDD_bonferroni=ifelse(Fisher_p_NDD_bonferroni>1,1,Fisher_p_NDD_bonferroni)) %>%\n",
                "  dplyr::select(class,tissue,num_expressed,num_unexpressed,Fisher_p_controls,Fisher_p_NDD,Fisher_p_controls_bonferroni,Fisher_p_NDD_bonferroni) %>%\n",
                "  data.table()\n",
                "\n",
                "colnames(supp_data_s10) <- c('Class','GTEx_Tissue','Num_Expressed Genes','Num_Unexpressed_Genes',\n",
                "                             'Fisher_P_Controls','Fisher_P_NDD','Fisher_P_Controls_Bonferroni',\n",
                "                             'Fisher_P_NDD_Bonferroni')\n",
                "\n",
                "write.table(supp_data_s10,'/results/supp/Supplementary_Data_S10_Proportion_hotspot_genes_expressed_across_tissues_all.txt',\n",
                "            col.names=TRUE,row.names=FALSE,quote=FALSE,sep=\"\\t\")\n",
                "\n",
                "#Supplementary data S11: Proportion of hotspot genes expressed across tissues, PF00520 domain-containing genes\n",
                "prop_expressed <- gene_tpm %>%\n",
                "  filter(contains_ion_channel==TRUE) %>%\n",
                "  group_by(tissue,class) %>%\n",
                "  summarise(\n",
                "    num_expressed = sum(tpm > 1, na.rm=T),\n",
                "    num_unexpressed = sum(tpm <= 1, na.rm=T),\n",
                "    .groups = 'drop_last'\n",
                "  ) %>%\n",
                "  filter(class!='Proposed Novel Hotspot Genes') %>%\n",
                "  data.table()\n",
                "\n",
                "prop_expressed <- prop_expressed[,Fisher_p_controls:=getFisher('Hotspot Genes','Control Genes',tissue),by=tissue]\n",
                "prop_expressed <- prop_expressed[,Fisher_p_NDD:=getFisher('Hotspot Genes','NDD-Associated Genes',tissue),by=tissue]\n",
                "prop_expressed <- prop_expressed[,Fisher_p_controls_bonferroni:=Fisher_p_controls*54] #Bonferroni correct over 54 GTEx tissues\n",
                "prop_expressed <- prop_expressed[,Fisher_p_NDD_bonferroni:=Fisher_p_NDD*54] #Bonferroni correct over 54 GTEx tissues\n",
                "\n",
                "supp_data_s11 <- prop_expressed %>%\n",
                "  mutate(Fisher_p_controls_bonferroni=ifelse(Fisher_p_controls_bonferroni>1,1,Fisher_p_controls_bonferroni),\n",
                "         Fisher_p_NDD_bonferroni=ifelse(Fisher_p_NDD_bonferroni>1,1,Fisher_p_NDD_bonferroni)) %>%\n",
                "  dplyr::select(class,tissue,num_expressed,num_unexpressed,Fisher_p_controls,Fisher_p_NDD,Fisher_p_controls_bonferroni,Fisher_p_NDD_bonferroni) %>%\n",
                "  data.table()\n",
                "\n",
                "colnames(supp_data_s11) <- c('Class','GTEx_Tissue','Num_Expressed Genes','Num_Unexpressed_Genes',\n",
                "                             'Fisher_P_Controls','Fisher_P_NDD','Fisher_P_Controls_Bonferroni',\n",
                "                             'Fisher_P_NDD_Bonferroni')\n",
                "\n",
                "write.table(supp_data_s11,'/results/supp/Supplementary_Data_S11_Proportion_hotspot_genes_expressed_across_tissues_PF00520.txt',\n",
                "            col.names=TRUE,row.names=FALSE,quote=FALSE,sep=\"\\t\")\n",
                "\n",
                "#Supplementary data S14: Gene sets used in analysis\n",
                "ddg2p_genes <- ddg2p_in %>%\n",
                "  pull(`gene symbol`) %>%\n",
                "  unique()\n",
                "  \n",
                "omim_genes <- omim_in %>%\n",
                "  separate_rows(`Gene Symbols`,sep=', ') %>%\n",
                "  filter(Phenotypes!='') %>%\n",
                "  pull(`Gene Symbols`) %>%\n",
                "  unique()\n",
                "\n",
                "constraint_genes <- constraint %>%\n",
                "  pull(gene) %>%\n",
                "  unique()\n",
                "\n",
                "supp_data_s14 <- gene_tpm %>%\n",
                "  mutate(pf00520 = contains_ion_channel) %>%\n",
                "  dplyr::select(hgnc,ensembl_versioned,class,pf00520) %>%\n",
                "  mutate(ddg2p=ifelse(hgnc %in% ddg2p_genes,TRUE,FALSE),\n",
                "         omim=ifelse(hgnc %in% omim_genes,TRUE,FALSE),\n",
                "         constraint=ifelse(hgnc %in% constraint_genes,TRUE,FALSE)) %>%\n",
                "  unique() %>%\n",
                "  data.table()\n",
                "\n",
                "write.table(supp_data_s14,'/results/supp/Supplementary_Data_S14_Gene_sets_all.txt',\n",
                "            col.names=TRUE,row.names=FALSE,quote=FALSE,sep=\"\\t\")\n",
                "\n",
                "#Supplementary data S15: PF00520 domain-containing genes used in analysis\n",
                "supp_data_s15 <- supp_data_s14 %>%\n",
                "  filter(pf00520==TRUE) %>%\n",
                "  data.table()\n",
                "\n",
                "write.table(supp_data_s15,'/results/supp/Supplementary_Data_S15_Gene_sets_PF00520.txt',\n",
                "            col.names=TRUE,row.names=FALSE,quote=FALSE,sep=\"\\t\")\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "### Supplementary tables\n",
                "\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Supplementary table 4: Hotspot genes are enriched for gain-of-function mutation consequences in DDG2P\n",
                "nonhap_list <- c('dominant negative','gain of function','activating','increased gene dosage')\n",
                "\n",
                "hotspot_ddg2p_activating_genes <- length(unique(ddg2p_in[`gene symbol` %in% hotspot_genes & `mutation consequence` %in% nonhap_list,`gene symbol`]))\n",
                "hotspot_ddg2p_genes <- length(unique(ddg2p_in[`gene symbol` %in% hotspot_genes,`gene symbol`])) - hotspot_ddg2p_activating_genes\n",
                "ddg2p_activating_genes <- length(unique(ddg2p_in[`mutation consequence` %in% nonhap_list,`gene symbol`])) - hotspot_ddg2p_activating_genes\n",
                "ddg2p_genes <- length(unique(ddg2p_in[,`gene symbol`])) - hotspot_ddg2p_genes - ddg2p_activating_genes\n",
                "\n",
                "supp_table_4 <- rbind(c(hotspot_ddg2p_activating_genes,hotspot_ddg2p_genes),\n",
                "                   c(ddg2p_activating_genes,ddg2p_genes))\n",
                "colnames(supp_table_4) <- c('Gain of Function Mutation Consequence','Other Mutation Consequence')\n",
                "row.names(supp_table_4) <- c('Hotspot Genes in DDG2P','Other DDG2P Genes')\n",
                "\n",
                "knitr::kable(supp_table_4)\n",
                "fisher.test(supp_table_4)\n",
                "\n",
                "#Supplementary table 5: NDD-associated genes have higher levels of constitutive expression than control genes\n",
                "constitutively_expressed_counts <- gene_tpm %>%\n",
                "  group_by(ensembl_versioned) %>%\n",
                "  mutate(expressed = ifelse(all(tpm < 1),'Unexpressed','Expressed')) %>%\n",
                "  mutate(constitutively_expressed = ifelse(all(tpm >= 1),'Constitutively Expressed','Not Constitutively Expressed')) %>%\n",
                "  ungroup() %>%\n",
                "  dplyr::select(ensembl_versioned,class,expressed,constitutively_expressed) %>%\n",
                "  filter(class=='Control Genes' | class=='NDD-Associated Genes') %>%\n",
                "  unique() %>%\n",
                "  count(class,expressed,constitutively_expressed) %>%\n",
                "  group_by(class,constitutively_expressed) %>%\n",
                "  summarise(n = sum(n),\n",
                "            .groups = 'keep') %>%\n",
                "  data.table()\n",
                "\n",
                "supp_table_5 <- rbind(constitutively_expressed_counts[class=='Control Genes',n],\n",
                "                      constitutively_expressed_counts[class=='NDD-Associated Genes',n])\n",
                "colnames(supp_table_5) <- c('Constitutively Expressed','Not Constitutively Expressed')\n",
                "row.names(supp_table_5) <- c('Control Genes','NDD-Associated Genes')\n",
                "\n",
                "knitr::kable(supp_table_5)\n",
                "fisher.test(supp_table_5)\n"
            ]
        },
        {
            "cell_type": "markdown",
            "metadata": {},
            "source": [
                "### Supplementary figures\n",
                "\n"
            ]
        },
        {
            "cell_type": "code",
            "execution_count": null,
            "metadata": {},
            "outputs": [],
            "source": [
                "#Supplementary figure 1: A significant proportion of hotspot genes have evidence of regional missense constraint compared to control and NDD-associated genes\n",
                "total_genes_rmc = length(unique(mpc_in[,gene]))\n",
                "dd_consensus_with_rmc = length(unique(mpc_in[gene %in% dd_genes,gene]))\n",
                "hotspot_with_rmc = length(unique(mpc_in[gene %in% hotspot_genes,gene]))\n",
                "novel_hotspot_with_rmc = length(unique(mpc_in[gene %in% novel_hotspot_genes,gene]))\n",
                "control_genes_with_rmc = total_genes_rmc - (dd_consensus_with_rmc + hotspot_with_rmc + novel_hotspot_with_rmc)\n",
                "\n",
                "hotspot_counts <- c(hotspot_with_rmc,length(hotspot_genes))\n",
                "dd_consensus_counts <- c(dd_consensus_with_rmc,length(dd_genes))\n",
                "control_counts = c(control_genes_with_rmc,length(unique(gene_tpm[class=='Control Genes',hgnc])))\n",
                "\n",
                "supp_fig1 <- data.table(rmc_counts=c(novel_hotspot_with_rmc,hotspot_with_rmc,dd_consensus_with_rmc,control_genes_with_rmc),\n",
                "                     total_counts=c(length(novel_hotspot_genes),length(hotspot_genes),length(dd_genes),length(unique(gene_tpm[class=='Control Genes',hgnc]))),\n",
                "                     class=c('Proposed Novel Hotspot Genes','Hotspot Genes','NDD-Associated Genes','Control Genes'))\n",
                "\n",
                "supp_fig1 <- supp_fig1 %>%\n",
                "  mutate(prop_rmc = rmc_counts/total_counts,\n",
                "         prop_other = 1 - prop_rmc) %>%\n",
                "  reshape2::melt(id.vars='class') %>%\n",
                "  data.table()\n",
                "\n",
                "supp_fig1_x_labels = c('Control Genes' = 'Control Genes',\n",
                "                    'Hotspot Genes' = 'Hotspot Genes',\n",
                "                    'NDD-Associated Genes' = 'NDD-Associated Genes',\n",
                "                    'Proposed Novel Hotspot Genes' = 'Proposed Novel\\nHotspot Genes')\n",
                "supp_fig1_labels = c('Genes without Evidence of Regional Missense Constraint','Genes with Evidence of Regional Missense Constraint')\n",
                "supp_fig1_fill = c('darkgrey','tomato4')\n",
                "\n",
                "ggplot(supp_c[variable=='prop_rmc' | variable=='prop_other'],\n",
                "       aes(x=class,y=value,fill=factor(variable,levels=c('prop_other','prop_rmc')))) +\n",
                "  geom_col() +\n",
                "  theme_pubr() +\n",
                "  theme(legend.position = 'top',\n",
                "        axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=14), \n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  scale_fill_manual(values=supp_fig1_fill,labels=supp_fig1_labels) +\n",
                "  scale_x_discrete(labels=supp_fig1_x_labels) +\n",
                "  xlab('') +\n",
                "  ylab('Proportion of Genes')\n",
                "\n",
                "ggsave('~/results/supp/Supplementary_Figure_1.svg',plot=last_plot(),device='svg',dpi=300,width=10,height=8)\n",
                "ggsave('~/results/supp/Supplementary_Figure_1.png',plot=last_plot(),device='png',dpi=300,width=10,height=8)\n",
                "\n",
                "fisher.test(rbind(hotspot_counts,control_counts))\n",
                "fisher.test(rbind(hotspot_counts,dd_consensus_counts))\n",
                "fisher.test(rbind(dd_consensus_counts,control_counts))\n",
                "\n",
                "#Supplementary figure 2: Most mutation hotspots are more constrained for missense variation than their global missense constraint score would suggest\n",
                "ggplot(fig_4b,aes(x=value,y=obs_exp,colour=class.x)) +\n",
                "  geom_point(size=7) +\n",
                "  geom_abline(slope=1,intercept=0,colour='black',linetype='dashed') +\n",
                "  theme_pubr() +\n",
                "  theme(axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=16),\n",
                "        legend.position = 'top',\n",
                "        legend.text = element_text(size=14),\n",
                "        legend.title = element_blank()) +\n",
                "  scale_colour_manual(values=c('skyblue3','darkgoldenrod','darkgreen','tomato4')) +\n",
                "  xlim(0,1) +\n",
                "  ylim(0,1) + \n",
                "  xlab('Global missense constraint') +\n",
                "  ylab('Regional missense constraint at hotspots')\n",
                "\n",
                "ggsave('~/results/supp/Supplementary_Figure_2.svg',plot=last_plot(),device='svg',dpi=300,width=10,height=8)\n",
                "ggsave('~/results/supp/Supplementary_Figure_2.png',plot=last_plot(),device='png',dpi=300,width=10,height=8)\n",
                "\n",
                "#Supplementary figure 3: A higher proportion of hotspot genes are expressed in brain than NDD-associated or control genes\n",
                "median_gene_tpm <- gene_tpm %>%\n",
                "  mutate(tissue_group = ifelse(tissue %in% brain_tissues,'Brain','Other Tissues')) %>%\n",
                "  group_by(ensembl_versioned,tissue_group) %>%\n",
                "  mutate(median_tpm_tissuegroup = median(tpm)) %>%\n",
                "  dplyr::select(-c('tpm','tissue')) %>%\n",
                "  unique() %>%\n",
                "  reshape2::dcast(formula=hgnc + class + ensembl_versioned + contains_ion_channel ~ tissue_group, value.var='median_tpm_tissuegroup') %>%\n",
                "  data.table()\n",
                "\n",
                "supp_fig3 <- median_gene_tpm %>%\n",
                "  group_by(class) %>%\n",
                "  summarise(\n",
                "    num_brain = sum(Brain > `Other Tissues`, na.rm=T),\n",
                "    num_other_tissues = sum(`Other Tissues` > Brain, na.rm=T),\n",
                "    total = n(),\n",
                "    num_unexpressed = total - (num_brain + num_other_tissues),\n",
                "    prop_brain = sum(Brain > `Other Tissues`, na.rm=T) / n(),\n",
                "    prop_other = sum(`Other Tissues` > Brain, na.rm=T) / n(),\n",
                "    prop_unexpressed = sum(Brain == `Other Tissues`, na.rm=T) / n(),\n",
                "    .groups='drop_last'\n",
                "  ) %>%\n",
                "  reshape2::melt(id.vars=c('class')) %>%\n",
                "  data.table()\n",
                "\n",
                "supp_fig3_x_labels = c('Control Genes','NDD-Associated Genes','Hotspot Genes','Proposed Novel\\nHotspot Genes')\n",
                "supp_fig3_fill = c('prop_unexpressed' = 'darkgrey',\n",
                "         'prop_other' = 'steelblue3',\n",
                "         'prop_brain' = 'steelblue4')\n",
                "supp_fig3_levels = c('prop_unexpressed','prop_other','prop_brain')\n",
                "supp_fig3_class_levels = c('Control Genes','NDD-Associated Genes','Hotspot Genes','Proposed Novel Hotspot Genes')\n",
                "supp_fig3_fill_labels = c('prop_unexpressed' = 'Unexpressed',\n",
                "                'prop_other' = 'Other Tissues TPM > Brain TPM',\n",
                "                'prop_brain' = 'Brain TPM > Other Tissues TPM')\n",
                "\n",
                "ggplot(supp_fig3[variable %in% c('prop_brain','prop_other','prop_unexpressed')],\n",
                "       aes(x=factor(class,levels=supp_fig3_class_levels),\n",
                "           y=value,fill=factor(variable,levels=supp_fig3_levels))) + \n",
                "  geom_col(colour='black',alpha=0.8) +\n",
                "  theme_pubr() +\n",
                "  theme(legend.position = 'top',\n",
                "        axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=14), \n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  scale_fill_manual(labels=supp_fig3_fill_labels,values=supp_fig3_fill) +\n",
                "  scale_x_discrete(labels=supp_fig3_x_labels) +\n",
                "  xlab('') +\n",
                "  ylab('Proportion of Genes')\n",
                "\n",
                "ggsave('~/results/supp/Supplementary_Figure_3.svg',plot=last_plot(),device='svg',dpi=300,width=10,height=8)\n",
                "ggsave('~/results/supp/Supplementary_Figure_3.png',plot=last_plot(),device='png',dpi=300,width=10,height=8)\n",
                "\n",
                "hotspot_counts <- c(supp_fig3[class=='Hotspot Genes' & variable=='num_brain',value],\n",
                "                    supp_fig3[class=='Hotspot Genes' & variable=='num_other_tissues',value])\n",
                "dd_counts <- c(supp_fig3[class=='NDD-Associated Genes' & variable=='num_brain',value],\n",
                "               supp_fig3[class=='NDD-Associated Genes' & variable=='num_other_tissues',value])\n",
                "control_counts <- c(supp_fig3[class=='Control Genes' & variable=='num_brain',value],\n",
                "                    supp_fig3[class=='Control Genes' & variable=='num_other_tissues',value])\n",
                "\n",
                "fisher.test(rbind(hotspot_counts,control_counts))\n",
                "fisher.test(rbind(hotspot_counts,dd_counts))\n",
                "\n",
                "#Supplementary figure 4: Proportion of hotspot genes expressed across tissues compared to PF00520 domain-containing NDD-associated genes and PF00520 domain-containing control genes\n",
                "supp_fig4 <- gene_tpm %>%\n",
                "  filter(contains_ion_channel==TRUE) %>%\n",
                "  filter(class!='Proposed Novel Hotspot Genes') %>%\n",
                "  group_by(tissue,class) %>%\n",
                "  summarise(\n",
                "    prop_expressed = sum(tpm > 1, na.rm=T) / n(),\n",
                "    .groups = 'keep'\n",
                "  ) %>%\n",
                "  data.table()\n",
                "\n",
                "supp_fig4_colour <- c('Control Genes' = 'skyblue3',\n",
                "                      'NDD-Associated Genes' = 'darkgreen',\n",
                "                      'Hotspot Genes' = 'darkgoldenrod')\n",
                "\n",
                "supp_fig4_labels <- c('Control Genes' = 'PF00520 Domain-Containing Control Genes',\n",
                "                      'NDD-Associated Genes' = 'PF00520 Domain-Containing NDD-Associated Genes',\n",
                "                      'Hotspot Genes' = 'Hotspot Genes')\n",
                "\n",
                "ggplot(supp_fig4,\n",
                "       aes(x=tissue,y=prop_expressed,colour=class)) +\n",
                "  geom_point(size=4) +\n",
                "  theme_pubr() +\n",
                "  theme(axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=16), \n",
                "        axis.text.x = element_text(angle = 90, \n",
                "                                   vjust = 0.5, \n",
                "                                   hjust=1),\n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  scale_colour_manual(values=supp_fig4_colour,labels=supp_fig4_labels) +\n",
                "  ylim(0,1) +\n",
                "  xlab('') +\n",
                "  ylab('Proportion expressed genes (TPM > 1)')\n",
                "\n",
                "ggsave('~/results/supp/Supplementary_Figure_4.svg',plot=last_plot(),device='svg',dpi=300,width=14,height=10)\n",
                "ggsave('~/results/supp/Supplementary_Figure_4.png',plot=last_plot(),device='png',dpi=300,width=14,height=10)\n",
                "\n",
                "#Supplementary figure 5: A higher proportion of hotspot genes are expressed in brain than PF00520 domain-containing control genes\n",
                "supp_fig5 <- median_gene_tpm %>%\n",
                "  filter(contains_ion_channel==TRUE) %>%\n",
                "  group_by(class) %>%\n",
                "  summarise(\n",
                "    num_brain = sum(Brain > `Other Tissues`, na.rm=T),\n",
                "    num_other_tissues = sum(`Other Tissues` > Brain, na.rm=T),\n",
                "    total = n(),\n",
                "    num_unexpressed = total - (num_brain + num_other_tissues),\n",
                "    prop_brain = sum(Brain > `Other Tissues`, na.rm=T) / n(),\n",
                "    prop_other = sum(`Other Tissues` > Brain, na.rm=T) / n(),\n",
                "    prop_unexpressed = sum(Brain == `Other Tissues`, na.rm=T) / n(),\n",
                "    .groups='drop_last'\n",
                "  ) %>%\n",
                "  reshape2::melt(id.vars=c('class')) %>%\n",
                "  data.table()\n",
                "\n",
                "supp_fig5_x_labels = c('PF00520 Domain-\\nContaining\\nControl Genes',\n",
                "                   'PF00520 Domain-\\nContaining\\nNDD-Associated Genes',\n",
                "                   'Hotspot Genes','Proposed Novel\\nHotspot Genes')\n",
                "supp_fig5_fill = c('prop_unexpressed' = 'darkgrey',\n",
                "         'prop_other' = 'steelblue3',\n",
                "         'prop_brain' = 'steelblue4')\n",
                "supp_fig5_levels = c('prop_unexpressed','prop_other','prop_brain')\n",
                "supp_fig5_class_levels = c('Control Genes','NDD-Associated Genes','Hotspot Genes','Proposed Novel Hotspot Genes')\n",
                "supp_fig5_fill_labels = c('prop_unexpressed' = 'Unexpressed',\n",
                "                'prop_other' = 'Other Tissues TPM > Brain TPM',\n",
                "                'prop_brain' = 'Brain TPM > Other Tissues TPM')\n",
                "\n",
                "ggplot(supp_fig5[variable %in% c('prop_brain','prop_other','prop_unexpressed')],\n",
                "       aes(x=factor(class,levels=supp_fig5_class_levels),\n",
                "           y=value,fill=factor(variable,levels=supp_fig5_levels))) + \n",
                "  geom_col(colour='black',alpha=0.8) +\n",
                "  theme_pubr() +\n",
                "  theme(legend.position = 'top',\n",
                "        axis.text = element_text(size=14), \n",
                "        axis.title = element_text(size=14), \n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  scale_fill_manual(labels=supp_fig5_fill_labels,values=supp_fig5_fill) +\n",
                "  scale_x_discrete(labels=supp_fig5_x_labels) +\n",
                "  xlab('') +\n",
                "  ylab('Proportion of Genes')\n",
                "\n",
                "ggsave('~/results/supp/Supplementary_Figure_5.svg',plot=last_plot(),device='svg',dpi=300,width=10,height=8)\n",
                "ggsave('~/results/supp/Supplementary_Figure_5.png',plot=last_plot(),device='png',dpi=300,width=10,height=8)\n",
                "\n",
                "hotspot_counts_PF00520 <- c(supp_fig5[class=='Hotspot Genes' & variable=='num_brain',value],\n",
                "                            supp_fig5[class=='Hotspot Genes' & variable=='num_other_tissues',value])\n",
                "dd_counts_PF00520 <- c(supp_fig5[class=='NDD-Associated Genes' & variable=='num_brain',value],\n",
                "                       supp_fig5[class=='NDD-Associated Genes' & variable=='num_other_tissues',value])\n",
                "control_counts_PF00520 <- c(supp_fig5[class=='Control Genes' & variable=='num_brain',value],\n",
                "                            supp_fig5[class=='Control Genes' & variable=='num_other_tissues',value])\n",
                "\n",
                "fisher.test(rbind(hotspot_counts_PF00520,control_counts_PF00520))\n",
                "fisher.test(rbind(hotspot_counts_PF00520,dd_counts_PF00520))\n",
                "\n",
                "#Supplementary figure 6: TPM differences between hotspot, NDD-associated, and control genes in brain and other tissues\n",
                "a <- ggplot(median_gene_tpm[class!='Proposed Novel Hotspot Genes'],\n",
                "            aes(x=factor(class,levels=c('Control Genes','NDD-Associated Genes','Hotspot Genes')),y=log10(Brain),fill=class)) +\n",
                "  geom_boxplot(outlier.colour = NA,alpha=0.8) +\n",
                "  theme_pubr() +\n",
                "  theme(axis.text = element_text(size=14),\n",
                "        axis.title = element_text(size=16), \n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  scale_fill_manual(values=c('skyblue3','darkgreen','darkgoldenrod')) +\n",
                "  ggsignif::geom_signif(comparisons = list(c('NDD-Associated Genes','Hotspot Genes'),\n",
                "                                           c('Control Genes','NDD-Associated Genes'),\n",
                "                                           c('Control Genes','Hotspot Genes')),\n",
                "                        map_signif_level = FALSE,\n",
                "                        y_position = c(3.5,3.8,4.2),\n",
                "                        colour='black') +\n",
                "  xlab('') +\n",
                "  ylab('Log10 Brain TPM')\n",
                "\n",
                "b <- ggplot(median_gene_tpm[class!='Proposed Novel Hotspot Genes'],\n",
                "            aes(x=factor(class,levels=c('Control Genes','NDD-Associated Genes','Hotspot Genes')),y=log10(`Other Tissues`),fill=class)) +\n",
                "  geom_boxplot(outlier.colour = NA,alpha=0.8) +\n",
                "  theme_pubr() +\n",
                "  theme(axis.text = element_text(size=14),\n",
                "        axis.title = element_text(size=16), \n",
                "        legend.title = element_blank(), \n",
                "        legend.text = element_text(size=14)) +\n",
                "  scale_fill_manual(values=c('skyblue3','darkgreen','darkgoldenrod')) +\n",
                "  ggsignif::geom_signif(comparisons = list(c('NDD-Associated Genes','Hotspot Genes'),\n",
                "                                           c('Control Genes','NDD-Associated Genes'),\n",
                "                                           c('Control Genes','Hotspot Genes')),\n",
                "                        map_signif_level = FALSE,\n",
                "                        y_position = c(3.5,3.8,4.2),\n",
                "                        colour='black') +\n",
                "  xlab('') +\n",
                "  ylab('Log10 Other Tissues TPM')\n",
                "  \n",
                "plot_grid(a,b,labels='AUTO',label_size=16,nrow=1)\n",
                "\n",
                "ggsave('~/results/supp/Supplementary_Figure_6.svg',plot=last_plot(),device='svg',dpi=300,width=14,height=8)\n",
                "ggsave('~/results/supp/Supplementary_Figure_6.png',plot=last_plot(),device='png',dpi=300,width=14,height=8)\n"
            ]
        }
    ],
    "metadata": {
        "anaconda-cloud": "",
        "kernelspec": {
            "display_name": "R",
            "langauge": "R",
            "name": "ir"
        },
        "language_info": {
            "codemirror_mode": "r",
            "file_extension": ".r",
            "mimetype": "text/x-r-source",
            "name": "R",
            "pygments_lexer": "r",
            "version": "3.4.1"
        }
    },
    "nbformat": 4,
    "nbformat_minor": 1
}
